<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>IBDMDB — 3-Omic Joint-RPCA (MGX + MTX + 16S)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="ibdmdb_3omic_jointrpca_files/libs/clipboard/clipboard.min.js"></script>
<script src="ibdmdb_3omic_jointrpca_files/libs/quarto-html/quarto.js"></script>
<script src="ibdmdb_3omic_jointrpca_files/libs/quarto-html/popper.min.js"></script>
<script src="ibdmdb_3omic_jointrpca_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ibdmdb_3omic_jointrpca_files/libs/quarto-html/anchor.min.js"></script>
<link href="ibdmdb_3omic_jointrpca_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ibdmdb_3omic_jointrpca_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ibdmdb_3omic_jointrpca_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ibdmdb_3omic_jointrpca_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ibdmdb_3omic_jointrpca_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">IBDMDB — 3-Omic Joint-RPCA (MGX + MTX + 16S)</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">warn =</span> <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">#load user-defined functions &amp; packages</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/dependencies.R"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointRPCA.R"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointRPCAuniversal.R"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointOptspaceHelper.R"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointOptspaceSolve.R"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/optspaceHelper.R"</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/transformHelper.R"</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/transform.R"</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/maskValueOnly.R"</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/rpcaTableProcessing.R"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointRPCAutils.R"</span>)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co">#paths</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>f_16s  <span class="ot">&lt;-</span> <span class="st">"data_ibdmdb_raw/taxonomic_profiles_16s_new.tsv"</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>f_mgx  <span class="ot">&lt;-</span> <span class="st">"data_ibdmdb_raw/taxonomic_profiles_mgx_new.tsv"</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>f_mtx  <span class="ot">&lt;-</span> <span class="st">"data_ibdmdb_raw/taxonomic_profiles_mtx_new.tsv"</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>f_meta <span class="ot">&lt;-</span> <span class="st">"data_ibdmdb_raw/hmp2_metadata_2018-08-20.csv"</span>  </span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="co">#helpers</span></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>read_ibdmdb_tsv <span class="ot">&lt;-</span> <span class="cf">function</span>(path) {</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stopifnot</span>(<span class="fu">file.exists</span>(path))</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>  first <span class="ot">&lt;-</span> <span class="fu">readLines</span>(path, <span class="at">n =</span> 200L, <span class="at">warn =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>  comment_idx <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">grepl</span>(<span class="st">"^#"</span>, first))</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(comment_idx) <span class="sc">==</span> 0L) <span class="fu">stop</span>(<span class="st">"No commented header line found in: "</span>, path)</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>  header_line <span class="ot">&lt;-</span> first[<span class="fu">max</span>(comment_idx)]</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>  header_line <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"^#</span><span class="sc">\\</span><span class="st">s*"</span>, <span class="st">""</span>, header_line)</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>  header_line <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"^\ufeff"</span>, <span class="st">""</span>, header_line)</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>  header_vec  <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(header_line, <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">fixed =</span> <span class="cn">TRUE</span>)[[<span class="dv">1</span>]]</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>  header_vec  <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">'^"|"$'</span>, <span class="st">""</span>, header_vec)</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>  dt <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(path, <span class="at">skip =</span> <span class="fu">length</span>(comment_idx), <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">ncol</span>(dt) <span class="sc">!=</span> <span class="fu">length</span>(header_vec)) {</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="fu">sprintf</span>(<span class="st">"Header columns (%d) != data columns (%d) in %s"</span>,</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">length</span>(header_vec), <span class="fu">ncol</span>(dt), path))</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>  data.table<span class="sc">::</span><span class="fu">setnames</span>(dt, header_vec)</span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>  dt</span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>to_matrix <span class="ot">&lt;-</span> <span class="cf">function</span>(dt) {</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a>  rn <span class="ot">&lt;-</span> dt[[<span class="dv">1</span>]]</span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(dt[, <span class="sc">-</span><span class="dv">1</span>, <span class="at">with =</span> <span class="cn">FALSE</span>])</span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(mat) <span class="ot">&lt;-</span> rn</span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">storage.mode</span>(mat) <span class="ot">&lt;-</span> <span class="st">"numeric"</span></span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>  mat[<span class="fu">is.na</span>(mat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>  mat</span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a>dedup_rownames <span class="ot">&lt;-</span> <span class="cf">function</span>(mat) {</span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stopifnot</span>(<span class="sc">!</span><span class="fu">is.null</span>(<span class="fu">rownames</span>(mat)))</span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a>  rn <span class="ot">&lt;-</span> <span class="fu">rownames</span>(mat)</span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a>  rn[rn <span class="sc">==</span> <span class="st">""</span> <span class="sc">|</span> <span class="fu">is.na</span>(rn)] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"feat_"</span>, <span class="fu">seq_len</span>(<span class="fu">sum</span>(rn <span class="sc">==</span> <span class="st">""</span> <span class="sc">|</span> <span class="fu">is.na</span>(rn))))</span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(mat) <span class="ot">&lt;-</span> <span class="fu">make.unique</span>(<span class="fu">as.character</span>(rn), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a>  mat</span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a>sanitize_matrix <span class="ot">&lt;-</span> <span class="cf">function</span>(mat) {</span>
<span id="cb1-61"><a href="#cb1-61" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(mat)</span>
<span id="cb1-62"><a href="#cb1-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">storage.mode</span>(mat) <span class="ot">&lt;-</span> <span class="st">"numeric"</span></span>
<span id="cb1-63"><a href="#cb1-63" aria-hidden="true" tabindex="-1"></a>  mat[<span class="sc">!</span><span class="fu">is.finite</span>(mat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb1-64"><a href="#cb1-64" aria-hidden="true" tabindex="-1"></a>  mat</span>
<span id="cb1-65"><a href="#cb1-65" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a><span class="st">`</span><span class="at">%||%</span><span class="st">`</span> <span class="ot">&lt;-</span> <span class="cf">function</span>(a, b) <span class="cf">if</span> (<span class="fu">is.null</span>(a)) b <span class="cf">else</span> a</span>
<span id="cb1-68"><a href="#cb1-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-69"><a href="#cb1-69" aria-hidden="true" tabindex="-1"></a><span class="co">#try to build IBD/non-IBD labels from metadata</span></span>
<span id="cb1-70"><a href="#cb1-70" aria-hidden="true" tabindex="-1"></a>make_groups_autodetect <span class="ot">&lt;-</span> <span class="cf">function</span>(meta_df, sample_ids, <span class="at">min_frac =</span> <span class="fl">0.01</span>, <span class="at">min_abs =</span> 10L) {</span>
<span id="cb1-71"><a href="#cb1-71" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample_id =</span> sample_ids, <span class="at">Group =</span> <span class="fu">factor</span>(<span class="cn">NA</span>, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"IBD"</span>, <span class="st">"non-IBD"</span>)))</span>
<span id="cb1-72"><a href="#cb1-72" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(meta_df) <span class="sc">||</span> <span class="sc">!</span><span class="fu">nrow</span>(meta_df)) <span class="fu">return</span>(out)</span>
<span id="cb1-73"><a href="#cb1-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-74"><a href="#cb1-74" aria-hidden="true" tabindex="-1"></a>  md <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(meta_df, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-75"><a href="#cb1-75" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(md) <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">names</span>(md)))</span>
<span id="cb1-76"><a href="#cb1-76" aria-hidden="true" tabindex="-1"></a>  sid <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(sample_ids)))</span>
<span id="cb1-77"><a href="#cb1-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-78"><a href="#cb1-78" aria-hidden="true" tabindex="-1"></a>  <span class="co">#choose the metadata column with the most overlap to our sample IDs</span></span>
<span id="cb1-79"><a href="#cb1-79" aria-hidden="true" tabindex="-1"></a>  overlaps <span class="ot">&lt;-</span> <span class="fu">sapply</span>(md, <span class="cf">function</span>(col) {</span>
<span id="cb1-80"><a href="#cb1-80" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(col)))</span>
<span id="cb1-81"><a href="#cb1-81" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(x) <span class="sc">&amp;</span> x <span class="sc">%in%</span> sid)</span>
<span id="cb1-82"><a href="#cb1-82" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb1-83"><a href="#cb1-83" aria-hidden="true" tabindex="-1"></a>  thresh <span class="ot">&lt;-</span> <span class="fu">max</span>(min_abs, <span class="fu">floor</span>(<span class="fu">length</span>(sid) <span class="sc">*</span> min_frac))</span>
<span id="cb1-84"><a href="#cb1-84" aria-hidden="true" tabindex="-1"></a>  max_ov <span class="ot">&lt;-</span> <span class="fu">suppressWarnings</span>(<span class="fu">max</span>(overlaps, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-85"><a href="#cb1-85" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.finite</span>(max_ov) <span class="sc">||</span> max_ov <span class="sc">&lt;</span> thresh) <span class="fu">return</span>(out)</span>
<span id="cb1-86"><a href="#cb1-86" aria-hidden="true" tabindex="-1"></a>  best <span class="ot">&lt;-</span> <span class="fu">names</span>(overlaps)[<span class="fu">which.max</span>(overlaps)]</span>
<span id="cb1-87"><a href="#cb1-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-88"><a href="#cb1-88" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="st">"diagnosis"</span> <span class="sc">%in%</span> <span class="fu">names</span>(md)) <span class="fu">return</span>(out)</span>
<span id="cb1-89"><a href="#cb1-89" aria-hidden="true" tabindex="-1"></a>  dx  <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(md<span class="sc">$</span>diagnosis)))</span>
<span id="cb1-90"><a href="#cb1-90" aria-hidden="true" tabindex="-1"></a>  grp <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">b(uc|cd|ibd)</span><span class="sc">\\</span><span class="st">b"</span>, dx), <span class="st">"IBD"</span>,</span>
<span id="cb1-91"><a href="#cb1-91" aria-hidden="true" tabindex="-1"></a>         <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"^</span><span class="sc">\\</span><span class="st">s*non"</span>, dx), <span class="st">"non-IBD"</span>, <span class="cn">NA_character_</span>))</span>
<span id="cb1-92"><a href="#cb1-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-93"><a href="#cb1-93" aria-hidden="true" tabindex="-1"></a>  md<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(md[[best]])))</span>
<span id="cb1-94"><a href="#cb1-94" aria-hidden="true" tabindex="-1"></a>  md<span class="sc">$</span>Group <span class="ot">&lt;-</span> <span class="fu">factor</span>(grp, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"IBD"</span>, <span class="st">"non-IBD"</span>))</span>
<span id="cb1-95"><a href="#cb1-95" aria-hidden="true" tabindex="-1"></a>  join_tbl <span class="ot">&lt;-</span> <span class="fu">unique</span>(md[, <span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"Group"</span>)])</span>
<span id="cb1-96"><a href="#cb1-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-97"><a href="#cb1-97" aria-hidden="true" tabindex="-1"></a>  j <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(<span class="fu">data.frame</span>(<span class="at">sample_id =</span> sid), join_tbl, <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb1-98"><a href="#cb1-98" aria-hidden="true" tabindex="-1"></a>  j<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> sample_ids</span>
<span id="cb1-99"><a href="#cb1-99" aria-hidden="true" tabindex="-1"></a>  j</span>
<span id="cb1-100"><a href="#cb1-100" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-101"><a href="#cb1-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-102"><a href="#cb1-102" aria-hidden="true" tabindex="-1"></a><span class="co">#evaluate U-scores with Wilcoxon / PERMANOVA / RF-AUROC</span></span>
<span id="cb1-103"><a href="#cb1-103" aria-hidden="true" tabindex="-1"></a>eval_scores <span class="ot">&lt;-</span> <span class="cf">function</span>(scores_df) {</span>
<span id="cb1-104"><a href="#cb1-104" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb1-105"><a href="#cb1-105" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span>(<span class="st">"Group"</span> <span class="sc">%in%</span> <span class="fu">names</span>(scores_df))) <span class="fu">return</span>(out)</span>
<span id="cb1-106"><a href="#cb1-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-107"><a href="#cb1-107" aria-hidden="true" tabindex="-1"></a>  comp_cols <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"^V</span><span class="sc">\\</span><span class="st">d+$"</span>, <span class="fu">names</span>(scores_df), <span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-108"><a href="#cb1-108" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">length</span>(comp_cols)) <span class="fu">return</span>(out)</span>
<span id="cb1-109"><a href="#cb1-109" aria-hidden="true" tabindex="-1"></a>  use_cols <span class="ot">&lt;-</span> comp_cols[<span class="fu">seq_len</span>(<span class="fu">min</span>(3L, <span class="fu">length</span>(comp_cols)))]</span>
<span id="cb1-110"><a href="#cb1-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-111"><a href="#cb1-111" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="st">"V1"</span> <span class="sc">%in%</span> <span class="fu">names</span>(scores_df)) {</span>
<span id="cb1-112"><a href="#cb1-112" aria-hidden="true" tabindex="-1"></a>    res_w1 <span class="ot">&lt;-</span> <span class="fu">try</span>(<span class="fu">wilcox.test</span>(scores_df<span class="sc">$</span>V1 <span class="sc">~</span> scores_df<span class="sc">$</span>Group, <span class="at">exact =</span> <span class="cn">FALSE</span>), <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-113"><a href="#cb1-113" aria-hidden="true" tabindex="-1"></a>    out<span class="sc">$</span>wilcox_PC1_p <span class="ot">&lt;-</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">inherits</span>(res_w1, <span class="st">"try-error"</span>)) res_w1<span class="sc">$</span>p.value <span class="cf">else</span> <span class="cn">NA_real_</span></span>
<span id="cb1-114"><a href="#cb1-114" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-115"><a href="#cb1-115" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="st">"V2"</span> <span class="sc">%in%</span> <span class="fu">names</span>(scores_df)) {</span>
<span id="cb1-116"><a href="#cb1-116" aria-hidden="true" tabindex="-1"></a>    res_w2 <span class="ot">&lt;-</span> <span class="fu">try</span>(<span class="fu">wilcox.test</span>(scores_df<span class="sc">$</span>V2 <span class="sc">~</span> scores_df<span class="sc">$</span>Group, <span class="at">exact =</span> <span class="cn">FALSE</span>), <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-117"><a href="#cb1-117" aria-hidden="true" tabindex="-1"></a>    out<span class="sc">$</span>wilcox_PC2_p <span class="ot">&lt;-</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">inherits</span>(res_w2, <span class="st">"try-error"</span>)) res_w2<span class="sc">$</span>p.value <span class="cf">else</span> <span class="cn">NA_real_</span></span>
<span id="cb1-118"><a href="#cb1-118" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-119"><a href="#cb1-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-120"><a href="#cb1-120" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(use_cols) <span class="sc">&gt;=</span> <span class="dv">2</span>) {</span>
<span id="cb1-121"><a href="#cb1-121" aria-hidden="true" tabindex="-1"></a>    perm_df <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(scores_df[, <span class="fu">c</span>(<span class="st">"Group"</span>, use_cols), <span class="at">drop =</span> <span class="cn">FALSE</span>])</span>
<span id="cb1-122"><a href="#cb1-122" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">nrow</span>(perm_df) <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">&amp;&amp;</span></span>
<span id="cb1-123"><a href="#cb1-123" aria-hidden="true" tabindex="-1"></a>        <span class="fu">is.factor</span>(perm_df<span class="sc">$</span>Group) <span class="sc">&amp;&amp;</span></span>
<span id="cb1-124"><a href="#cb1-124" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nlevels</span>(perm_df<span class="sc">$</span>Group) <span class="sc">&gt;=</span> <span class="dv">2</span> <span class="sc">&amp;&amp;</span></span>
<span id="cb1-125"><a href="#cb1-125" aria-hidden="true" tabindex="-1"></a>        <span class="fu">all</span>(<span class="fu">table</span>(perm_df<span class="sc">$</span>Group) <span class="sc">&gt;=</span> <span class="dv">3</span>)) {</span>
<span id="cb1-126"><a href="#cb1-126" aria-hidden="true" tabindex="-1"></a>      comp_mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(perm_df[, use_cols, <span class="at">drop =</span> <span class="cn">FALSE</span>])</span>
<span id="cb1-127"><a href="#cb1-127" aria-hidden="true" tabindex="-1"></a>      <span class="fu">colnames</span>(comp_mat) <span class="ot">&lt;-</span> use_cols</span>
<span id="cb1-128"><a href="#cb1-128" aria-hidden="true" tabindex="-1"></a>      perm <span class="ot">&lt;-</span> <span class="fu">try</span>(vegan<span class="sc">::</span><span class="fu">adonis2</span>(comp_mat <span class="sc">~</span> Group, <span class="at">data =</span> perm_df, <span class="at">method =</span> <span class="st">"euclidean"</span>), <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-129"><a href="#cb1-129" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="sc">!</span><span class="fu">inherits</span>(perm, <span class="st">"try-error"</span>)) {</span>
<span id="cb1-130"><a href="#cb1-130" aria-hidden="true" tabindex="-1"></a>        out<span class="sc">$</span>permanova_R2 <span class="ot">&lt;-</span> perm<span class="sc">$</span>R2[<span class="dv">1</span>]</span>
<span id="cb1-131"><a href="#cb1-131" aria-hidden="true" tabindex="-1"></a>        out<span class="sc">$</span>permanova_F  <span class="ot">&lt;-</span> perm<span class="sc">$</span>F[<span class="dv">1</span>]</span>
<span id="cb1-132"><a href="#cb1-132" aria-hidden="true" tabindex="-1"></a>        out<span class="sc">$</span>permanova_p  <span class="ot">&lt;-</span> perm<span class="sc">$</span><span class="st">`</span><span class="at">Pr(&gt;F)</span><span class="st">`</span>[<span class="dv">1</span>]</span>
<span id="cb1-133"><a href="#cb1-133" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb1-134"><a href="#cb1-134" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb1-135"><a href="#cb1-135" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-136"><a href="#cb1-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-137"><a href="#cb1-137" aria-hidden="true" tabindex="-1"></a>  rf_df <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(scores_df[, <span class="fu">c</span>(<span class="st">"Group"</span>, use_cols), <span class="at">drop =</span> <span class="cn">FALSE</span>])</span>
<span id="cb1-138"><a href="#cb1-138" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">nrow</span>(rf_df) <span class="sc">&amp;&amp;</span> <span class="fu">is.factor</span>(rf_df<span class="sc">$</span>Group) <span class="sc">&amp;&amp;</span> <span class="fu">nlevels</span>(rf_df<span class="sc">$</span>Group) <span class="sc">&gt;=</span> <span class="dv">2</span>) {</span>
<span id="cb1-139"><a href="#cb1-139" aria-hidden="true" tabindex="-1"></a>    cls_tab <span class="ot">&lt;-</span> <span class="fu">table</span>(rf_df<span class="sc">$</span>Group)</span>
<span id="cb1-140"><a href="#cb1-140" aria-hidden="true" tabindex="-1"></a>    wts <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="dv">1</span> <span class="sc">/</span> cls_tab); <span class="fu">names</span>(wts) <span class="ot">&lt;-</span> <span class="fu">names</span>(cls_tab)</span>
<span id="cb1-141"><a href="#cb1-141" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb1-142"><a href="#cb1-142" aria-hidden="true" tabindex="-1"></a>    rf_prob <span class="ot">&lt;-</span> ranger<span class="sc">::</span><span class="fu">ranger</span>(Group <span class="sc">~</span> ., <span class="at">data =</span> rf_df, <span class="at">num.trees =</span> <span class="dv">400</span>,</span>
<span id="cb1-143"><a href="#cb1-143" aria-hidden="true" tabindex="-1"></a>                              <span class="at">probability =</span> <span class="cn">TRUE</span>, <span class="at">class.weights =</span> wts, <span class="at">oob.error =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-144"><a href="#cb1-144" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="st">"IBD"</span> <span class="sc">%in%</span> <span class="fu">colnames</span>(rf_prob<span class="sc">$</span>predictions)) {</span>
<span id="cb1-145"><a href="#cb1-145" aria-hidden="true" tabindex="-1"></a>      p_ibd <span class="ot">&lt;-</span> rf_prob<span class="sc">$</span>predictions[, <span class="st">"IBD"</span>]</span>
<span id="cb1-146"><a href="#cb1-146" aria-hidden="true" tabindex="-1"></a>      roc_obj <span class="ot">&lt;-</span> pROC<span class="sc">::</span><span class="fu">roc</span>(rf_df<span class="sc">$</span>Group, p_ibd, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"non-IBD"</span>, <span class="st">"IBD"</span>))</span>
<span id="cb1-147"><a href="#cb1-147" aria-hidden="true" tabindex="-1"></a>      out<span class="sc">$</span>AUROC <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(pROC<span class="sc">::</span><span class="fu">auc</span>(roc_obj))</span>
<span id="cb1-148"><a href="#cb1-148" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb1-149"><a href="#cb1-149" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-150"><a href="#cb1-150" aria-hidden="true" tabindex="-1"></a>  out</span>
<span id="cb1-151"><a href="#cb1-151" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-152"><a href="#cb1-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-153"><a href="#cb1-153" aria-hidden="true" tabindex="-1"></a><span class="co">#helper: color by any metadata variable </span></span>
<span id="cb1-154"><a href="#cb1-154" aria-hidden="true" tabindex="-1"></a>color_by_meta <span class="ot">&lt;-</span> <span class="cf">function</span>(meta_df, sample_ids, <span class="at">id_cols =</span> <span class="fu">c</span>(<span class="st">"external id"</span>,<span class="st">"external.id"</span>,<span class="st">"external_id"</span>,</span>
<span id="cb1-155"><a href="#cb1-155" aria-hidden="true" tabindex="-1"></a>                                                           <span class="st">"sample id"</span>,<span class="st">"sample.id"</span>,<span class="st">"sample_id"</span>,</span>
<span id="cb1-156"><a href="#cb1-156" aria-hidden="true" tabindex="-1"></a>                                                           <span class="st">"mgx.sampleid"</span>,<span class="st">"mtx.sampleid"</span>,</span>
<span id="cb1-157"><a href="#cb1-157" aria-hidden="true" tabindex="-1"></a>                                                           <span class="st">"stool sample id:  tube a  (etoh)"</span>,</span>
<span id="cb1-158"><a href="#cb1-158" aria-hidden="true" tabindex="-1"></a>                                                           <span class="st">"sample id: tube b (no preservative)"</span>,</span>
<span id="cb1-159"><a href="#cb1-159" aria-hidden="true" tabindex="-1"></a>                                                           <span class="st">"stool_id"</span>,<span class="st">"wr id"</span>,<span class="st">"wr_id"</span>,<span class="st">"wrid"</span>),</span>
<span id="cb1-160"><a href="#cb1-160" aria-hidden="true" tabindex="-1"></a>                          value_col) {</span>
<span id="cb1-161"><a href="#cb1-161" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(meta_df) <span class="sc">||</span> <span class="sc">!</span><span class="fu">nrow</span>(meta_df) <span class="sc">||</span> <span class="fu">is.null</span>(value_col) <span class="sc">||</span> <span class="sc">!</span><span class="fu">tolower</span>(value_col) <span class="sc">%in%</span> <span class="fu">tolower</span>(<span class="fu">names</span>(meta_df))) {</span>
<span id="cb1-162"><a href="#cb1-162" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">data.frame</span>(<span class="at">sample_id =</span> sample_ids, <span class="at">Val =</span> <span class="cn">NA</span>))</span>
<span id="cb1-163"><a href="#cb1-163" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-164"><a href="#cb1-164" aria-hidden="true" tabindex="-1"></a>  m0 <span class="ot">&lt;-</span> meta_df</span>
<span id="cb1-165"><a href="#cb1-165" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(m0) <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">names</span>(m0)))</span>
<span id="cb1-166"><a href="#cb1-166" aria-hidden="true" tabindex="-1"></a>  id_cols <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">tolower</span>(id_cols), <span class="fu">names</span>(m0))</span>
<span id="cb1-167"><a href="#cb1-167" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">length</span>(id_cols)) id_cols <span class="ot">&lt;-</span> <span class="fu">names</span>(m0)</span>
<span id="cb1-168"><a href="#cb1-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-169"><a href="#cb1-169" aria-hidden="true" tabindex="-1"></a>  sid_norm <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(sample_ids))</span>
<span id="cb1-170"><a href="#cb1-170" aria-hidden="true" tabindex="-1"></a>  <span class="co">#choose the id col with max overlap</span></span>
<span id="cb1-171"><a href="#cb1-171" aria-hidden="true" tabindex="-1"></a>  best <span class="ot">&lt;-</span> id_cols[<span class="fu">which.max</span>(<span class="fu">sapply</span>(id_cols, <span class="cf">function</span>(c) {</span>
<span id="cb1-172"><a href="#cb1-172" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sum</span>(<span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(m0[[c]]))) <span class="sc">%in%</span> sid_norm)</span>
<span id="cb1-173"><a href="#cb1-173" aria-hidden="true" tabindex="-1"></a>  }))]</span>
<span id="cb1-174"><a href="#cb1-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-175"><a href="#cb1-175" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb1-176"><a href="#cb1-176" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_id_norm =</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(m0[[best]]))),</span>
<span id="cb1-177"><a href="#cb1-177" aria-hidden="true" tabindex="-1"></a>    <span class="at">Val =</span> m0[[<span class="fu">tolower</span>(value_col)]],</span>
<span id="cb1-178"><a href="#cb1-178" aria-hidden="true" tabindex="-1"></a>    <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span></span>
<span id="cb1-179"><a href="#cb1-179" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb1-180"><a href="#cb1-180" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">unique</span>(df[<span class="sc">!</span><span class="fu">is.na</span>(df<span class="sc">$</span>sample_id_norm) <span class="sc">&amp;</span> <span class="fu">nzchar</span>(df<span class="sc">$</span>sample_id_norm), ])</span>
<span id="cb1-181"><a href="#cb1-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-182"><a href="#cb1-182" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb1-183"><a href="#cb1-183" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(<span class="at">sample_id_norm =</span> sid_norm, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>),</span>
<span id="cb1-184"><a href="#cb1-184" aria-hidden="true" tabindex="-1"></a>    df, <span class="at">by =</span> <span class="st">"sample_id_norm"</span></span>
<span id="cb1-185"><a href="#cb1-185" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb1-186"><a href="#cb1-186" aria-hidden="true" tabindex="-1"></a>  out<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> sample_ids</span>
<span id="cb1-187"><a href="#cb1-187" aria-hidden="true" tabindex="-1"></a>  out<span class="sc">$</span>sample_id_norm <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb1-188"><a href="#cb1-188" aria-hidden="true" tabindex="-1"></a>  out</span>
<span id="cb1-189"><a href="#cb1-189" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-190"><a href="#cb1-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-191"><a href="#cb1-191" aria-hidden="true" tabindex="-1"></a><span class="co">#load data</span></span>
<span id="cb1-192"><a href="#cb1-192" aria-hidden="true" tabindex="-1"></a>dt_16s <span class="ot">&lt;-</span> <span class="fu">read_ibdmdb_tsv</span>(f_16s)</span>
<span id="cb1-193"><a href="#cb1-193" aria-hidden="true" tabindex="-1"></a>dt_mgx <span class="ot">&lt;-</span> <span class="fu">read_ibdmdb_tsv</span>(f_mgx)</span>
<span id="cb1-194"><a href="#cb1-194" aria-hidden="true" tabindex="-1"></a>dt_mtx <span class="ot">&lt;-</span> <span class="fu">read_ibdmdb_tsv</span>(f_mtx)</span>
<span id="cb1-195"><a href="#cb1-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-196"><a href="#cb1-196" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Dims (rows, cols):</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Dims (rows, cols):</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"16S:"</span>, <span class="fu">dim</span>(dt_16s), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>16S: 503 59 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"MGX:"</span>, <span class="fu">dim</span>(dt_mgx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MGX: 963 317 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"MTX:"</span>, <span class="fu">dim</span>(dt_mtx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MTX: 566 94 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>M_16s <span class="ot">&lt;-</span> <span class="fu">to_matrix</span>(dt_16s)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>M_mgx <span class="ot">&lt;-</span> <span class="fu">to_matrix</span>(dt_mgx)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>M_mtx <span class="ot">&lt;-</span> <span class="fu">to_matrix</span>(dt_mtx)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#strict 3-way intersection of sample IDs</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>samples_16s <span class="ot">&lt;-</span> <span class="fu">colnames</span>(M_16s)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>samples_mgx <span class="ot">&lt;-</span> <span class="fu">colnames</span>(M_mgx)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>samples_mtx <span class="ot">&lt;-</span> <span class="fu">colnames</span>(M_mtx)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>shared_3 <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(intersect, <span class="fu">list</span>(samples_16s, samples_mgx, samples_mtx))</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>shared_3 <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(shared_3[<span class="fu">nchar</span>(shared_3) <span class="sc">&gt;</span> <span class="dv">0</span>]))</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Strict 3-omic shared samples:"</span>, <span class="fu">length</span>(shared_3), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Strict 3-omic shared samples: 16 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">length</span>(shared_3) <span class="sc">&gt;=</span> <span class="dv">10</span>)  </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>X_16s <span class="ot">&lt;-</span> M_16s[, shared_3, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> M_mgx[, shared_3, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> M_mtx[, shared_3, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co">#per-view filtering</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>n_samp <span class="ot">&lt;-</span> <span class="fu">length</span>(shared_3)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>prev_16s <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fl">0.05</span> <span class="sc">*</span> n_samp) <span class="co">#5% prevalence for 16S</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>prev_mgx <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fl">0.05</span> <span class="sc">*</span> n_samp) <span class="co">#5% for MGX</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>prev_mtx <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fl">0.02</span> <span class="sc">*</span> n_samp) <span class="co">#2% for MTX </span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>keep_16s <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(X_16s <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> prev_16s</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>keep_mgx <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(X_mgx <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> prev_mgx</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>keep_mtx <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(X_mtx <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> prev_mtx</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>X_16s <span class="ot">&lt;-</span> X_16s[keep_16s, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> X_mgx[keep_mgx, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> X_mtx[keep_mtx, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a><span class="co">#drop any all-zero samples within a view</span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>keep_samp_16s <span class="ot">&lt;-</span> <span class="fu">colSums</span>(X_16s) <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>keep_samp_mgx <span class="ot">&lt;-</span> <span class="fu">colSums</span>(X_mgx) <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>keep_samp_mtx <span class="ot">&lt;-</span> <span class="fu">colSums</span>(X_mtx) <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">all</span>(keep_samp_16s)) <span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[16S] dropping %d all-zero samples"</span>, <span class="fu">sum</span>(<span class="sc">!</span>keep_samp_16s)))</span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">all</span>(keep_samp_mgx)) <span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[MGX] dropping %d all-zero samples"</span>, <span class="fu">sum</span>(<span class="sc">!</span>keep_samp_mgx)))</span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">all</span>(keep_samp_mtx)) <span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[MTX] dropping %d all-zero samples"</span>, <span class="fu">sum</span>(<span class="sc">!</span>keep_samp_mtx)))</span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a>X_16s <span class="ot">&lt;-</span> X_16s[, keep_samp_16s, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> X_mgx[, keep_samp_mgx, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> X_mtx[, keep_samp_mtx, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a><span class="co">#recompute strict shared samples after pruning</span></span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a>shared_final <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(intersect, <span class="fu">list</span>(<span class="fu">colnames</span>(X_16s), <span class="fu">colnames</span>(X_mgx), <span class="fu">colnames</span>(X_mtx)))</span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a>shared_final <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(shared_final))</span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">length</span>(shared_final) <span class="sc">&gt;=</span> <span class="dv">10</span>)</span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a>X_16s <span class="ot">&lt;-</span> X_16s[, shared_final, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> X_mgx[, shared_final, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> X_mtx[, shared_final, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-42"><a href="#cb11-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-43"><a href="#cb11-43" aria-hidden="true" tabindex="-1"></a><span class="co">#cap largest view by variance</span></span>
<span id="cb11-44"><a href="#cb11-44" aria-hidden="true" tabindex="-1"></a>max_features <span class="ot">&lt;-</span> <span class="dv">15000</span></span>
<span id="cb11-45"><a href="#cb11-45" aria-hidden="true" tabindex="-1"></a>cap_by_var <span class="ot">&lt;-</span> <span class="cf">function</span>(M, cap) {</span>
<span id="cb11-46"><a href="#cb11-46" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">nrow</span>(M) <span class="sc">&lt;=</span> cap) <span class="fu">return</span>(M)</span>
<span id="cb11-47"><a href="#cb11-47" aria-hidden="true" tabindex="-1"></a>  ord <span class="ot">&lt;-</span> <span class="fu">order</span>(matrixStats<span class="sc">::</span><span class="fu">rowVars</span>(M), <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-48"><a href="#cb11-48" aria-hidden="true" tabindex="-1"></a>  M[ord[<span class="fu">seq_len</span>(cap)], , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb11-49"><a href="#cb11-49" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-50"><a href="#cb11-50" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> <span class="fu">cap_by_var</span>(X_mtx, max_features)</span>
<span id="cb11-51"><a href="#cb11-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-52"><a href="#cb11-52" aria-hidden="true" tabindex="-1"></a><span class="co">#clean names / numerics</span></span>
<span id="cb11-53"><a href="#cb11-53" aria-hidden="true" tabindex="-1"></a>X_16s <span class="ot">&lt;-</span> <span class="fu">sanitize_matrix</span>(<span class="fu">dedup_rownames</span>(X_16s))</span>
<span id="cb11-54"><a href="#cb11-54" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> <span class="fu">sanitize_matrix</span>(<span class="fu">dedup_rownames</span>(X_mgx))</span>
<span id="cb11-55"><a href="#cb11-55" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> <span class="fu">sanitize_matrix</span>(<span class="fu">dedup_rownames</span>(X_mtx))</span>
<span id="cb11-56"><a href="#cb11-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-57"><a href="#cb11-57" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Remaining features after filtering:</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb11-58"><a href="#cb11-58" aria-hidden="true" tabindex="-1"></a>    <span class="st">"16S:"</span>, <span class="fu">nrow</span>(X_16s), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb11-59"><a href="#cb11-59" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MGX:"</span>, <span class="fu">nrow</span>(X_mgx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb11-60"><a href="#cb11-60" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MTX:"</span>, <span class="fu">nrow</span>(X_mtx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Remaining features after filtering:
 16S: 233 
 MGX: 462 
 MTX: 370 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Final dims — 16S:"</span>, <span class="fu">nrow</span>(X_16s), <span class="st">"x"</span>, <span class="fu">ncol</span>(X_16s),</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"| MGX:"</span>, <span class="fu">nrow</span>(X_mgx), <span class="st">"x"</span>, <span class="fu">ncol</span>(X_mgx),</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"| MTX:"</span>, <span class="fu">nrow</span>(X_mtx), <span class="st">"x"</span>, <span class="fu">ncol</span>(X_mtx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Final dims — 16S: 233 x 16 | MGX: 462 x 16 | MTX: 370 x 16 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">identical</span>(<span class="fu">colnames</span>(X_16s), <span class="fu">colnames</span>(X_mgx)))</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">identical</span>(<span class="fu">colnames</span>(X_mgx), <span class="fu">colnames</span>(X_mtx)))</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co">#build MAE (3 assays)</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>cd <span class="ot">&lt;-</span> S4Vectors<span class="sc">::</span><span class="fu">DataFrame</span>(<span class="at">row.names =</span> shared_final)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(X_16s, <span class="st">"rn"</span>) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(X_16s)</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(X_mgx, <span class="st">"rn"</span>) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(X_mgx)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(X_mtx, <span class="st">"rn"</span>) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(X_mtx)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co">#create SEs</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>se_MGX <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">SummarizedExperiment</span>(<span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> X_mgx), <span class="at">colData =</span> cd)</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>se_MTX <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">SummarizedExperiment</span>(<span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> X_mtx), <span class="at">colData =</span> cd)</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>se_S16 <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">SummarizedExperiment</span>(<span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> X_16s), <span class="at">colData =</span> cd)</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="co">#fixed order and syntactic key for 16S</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>mae_3 <span class="ot">&lt;-</span> MultiAssayExperiment<span class="sc">::</span><span class="fu">MultiAssayExperiment</span>(</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">experiments =</span> <span class="fu">list</span>(<span class="at">MGX =</span> se_MGX, <span class="at">MTX =</span> se_MTX, <span class="at">S16 =</span> se_S16)</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>mae_3 <span class="ot">&lt;-</span> MultiAssayExperiment<span class="sc">::</span><span class="fu">intersectColumns</span>(mae_3)</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">identical</span>(<span class="fu">rownames</span>(SummarizedExperiment<span class="sc">::</span><span class="fu">assay</span>(mae_3[[<span class="st">"MGX"</span>]], <span class="st">"counts"</span>)), <span class="fu">attr</span>(X_mgx, <span class="st">"rn"</span>)),</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>          <span class="fu">identical</span>(<span class="fu">rownames</span>(SummarizedExperiment<span class="sc">::</span><span class="fu">assay</span>(mae_3[[<span class="st">"MTX"</span>]], <span class="st">"counts"</span>)), <span class="fu">attr</span>(X_mtx, <span class="st">"rn"</span>)),</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>          <span class="fu">identical</span>(<span class="fu">rownames</span>(SummarizedExperiment<span class="sc">::</span><span class="fu">assay</span>(mae_3[[<span class="st">"S16"</span>]], <span class="st">"counts"</span>)), <span class="fu">attr</span>(X_16s, <span class="st">"rn"</span>)))</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="co">#choose rank k safely</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>per_view_min_dim <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">list</span>(X_16s, X_mgx, X_mtx), <span class="cf">function</span>(m) <span class="fu">min</span>(<span class="fu">nrow</span>(m), <span class="fu">ncol</span>(m)))</span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>k_max <span class="ot">&lt;-</span> <span class="fu">max</span>(1L, <span class="fu">min</span>(per_view_min_dim))</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">min</span>(3L, k_max)</span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[3-omic] per-view min dims = %s; using k = %d"</span>,</span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>                <span class="fu">paste</span>(per_view_min_dim, <span class="at">collapse =</span> <span class="st">", "</span>), k))</span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a><span class="co">#run Joint-RPCA </span></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>fit3 <span class="ot">&lt;-</span> <span class="fu">jointRPCAuniversal</span>(</span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> mae_3,</span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a>  <span class="at">n.components =</span> k,</span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>  <span class="at">max.iterations =</span> <span class="dv">1000</span>,</span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>  <span class="at">rclr.transform.tables =</span> <span class="cn">TRUE</span>,</span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.sample.count =</span> <span class="dv">1</span>,</span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.feature.count =</span> <span class="dv">0</span>,</span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.feature.frequency =</span> <span class="dv">0</span></span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a>U <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(fit3<span class="sc">$</span>ord.res<span class="sc">$</span>samples)</span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(U) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"V"</span>, <span class="fu">seq_len</span>(<span class="fu">ncol</span>(U)))</span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a>U<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">rownames</span>(U)</span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a>get_view <span class="ot">&lt;-</span> <span class="cf">function</span>(obj, keys) {</span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(obj)) <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.list</span>(obj)) { <span class="cf">for</span> (k <span class="cf">in</span> keys) <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(obj[[k]])) <span class="fu">return</span>(<span class="fu">as.data.frame</span>(obj[[k]])); <span class="fu">return</span>(<span class="cn">NULL</span>) }</span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.matrix</span>(obj) <span class="sc">||</span> <span class="fu">is.data.frame</span>(obj)) <span class="fu">return</span>(<span class="fu">as.data.frame</span>(obj))</span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a>  <span class="cn">NULL</span></span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-55"><a href="#cb15-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-56"><a href="#cb15-56" aria-hidden="true" tabindex="-1"></a>V_MGX <span class="ot">&lt;-</span> <span class="fu">get_view</span>(fit3<span class="sc">$</span>ord.res<span class="sc">$</span>features, <span class="fu">c</span>(<span class="st">"MGX"</span>, <span class="st">"view_MGX"</span>, <span class="st">"view_mgx"</span>))</span>
<span id="cb15-57"><a href="#cb15-57" aria-hidden="true" tabindex="-1"></a>V_MTX <span class="ot">&lt;-</span> <span class="fu">get_view</span>(fit3<span class="sc">$</span>ord.res<span class="sc">$</span>features, <span class="fu">c</span>(<span class="st">"MTX"</span>, <span class="st">"view_MTX"</span>, <span class="st">"view_mtx"</span>))</span>
<span id="cb15-58"><a href="#cb15-58" aria-hidden="true" tabindex="-1"></a>V_S16 <span class="ot">&lt;-</span> <span class="fu">get_view</span>(fit3<span class="sc">$</span>ord.res<span class="sc">$</span>features, <span class="fu">c</span>(<span class="st">"S16"</span>, <span class="st">"view_S16"</span>, <span class="st">"view_s16"</span>))</span>
<span id="cb15-59"><a href="#cb15-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-60"><a href="#cb15-60" aria-hidden="true" tabindex="-1"></a><span class="co">#basic ordination (unlabeled)</span></span>
<span id="cb15-61"><a href="#cb15-61" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(</span>
<span id="cb15-62"><a href="#cb15-62" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(U, ggplot2<span class="sc">::</span><span class="fu">aes</span>(V1, V2)) <span class="sc">+</span></span>
<span id="cb15-63"><a href="#cb15-63" aria-hidden="true" tabindex="-1"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.85</span>) <span class="sc">+</span></span>
<span id="cb15-64"><a href="#cb15-64" aria-hidden="true" tabindex="-1"></a>    ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">sprintf</span>(<span class="st">"Joint-RPCA (16S + MGX + MTX), k = %d — PC1 vs PC2"</span>, k),</span>
<span id="cb15-65"><a href="#cb15-65" aria-hidden="true" tabindex="-1"></a>                  <span class="at">x =</span> <span class="st">"PC1"</span>, <span class="at">y =</span> <span class="st">"PC2"</span>) <span class="sc">+</span></span>
<span id="cb15-66"><a href="#cb15-66" aria-hidden="true" tabindex="-1"></a>    ggplot2<span class="sc">::</span><span class="fu">theme_minimal</span>()</span>
<span id="cb15-67"><a href="#cb15-67" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ibdmdb_3omic_jointrpca_files/figure-html/setup-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#attach IBD/non-IBD from metadata </span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>meta_df <span class="ot">&lt;-</span> <span class="cf">if</span> (<span class="fu">file.exists</span>(f_meta)) <span class="fu">data.frame</span>(data.table<span class="sc">::</span><span class="fu">fread</span>(f_meta)) <span class="cf">else</span> <span class="cn">NULL</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>grp_df  <span class="ot">&lt;-</span> <span class="fu">make_groups_autodetect</span>(meta_df, U<span class="sc">$</span>sample_id)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">sort</span>(<span class="fu">table</span>(grp_df<span class="sc">$</span>Group, <span class="at">useNA =</span> <span class="st">"ifany"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
non-IBD     IBD 
      0      16 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>U2 <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(U, grp_df, <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>tab <span class="ot">&lt;-</span> <span class="fu">table</span>(U2<span class="sc">$</span>Group, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    IBD non-IBD 
     16       0 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#top loadings per view (PC1)</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>top_k <span class="ot">&lt;-</span> <span class="dv">15</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>show_top <span class="ot">&lt;-</span> <span class="cf">function</span>(V, label) {</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(V) <span class="sc">||</span> <span class="sc">!</span><span class="fu">ncol</span>(V)) <span class="fu">return</span>(<span class="fu">invisible</span>(<span class="cn">NULL</span>))</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  ord <span class="ot">&lt;-</span> <span class="fu">order</span>(V[,<span class="dv">1</span>], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  top_ix <span class="ot">&lt;-</span> <span class="fu">seq_len</span>(<span class="fu">min</span>(top_k, <span class="fu">length</span>(ord)))</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Top %s features on PC1:</span><span class="sc">\n</span><span class="st">"</span>, label))</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">data.frame</span>(</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature =</span> <span class="fu">rownames</span>(V)[ord][top_ix],</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">loading =</span> V[ord, <span class="dv">1</span>][top_ix]</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="fu">show_top</span>(V_S16, <span class="st">"S16"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Top S16 features on PC1:
                                                                                                                                    feature
1                                          k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Ruminococcaceae; g__; s__:307224
2                   k__Bacteria; p__Proteobacteria; c__Betaproteobacteria; o__Burkholderiales; f__Alcaligenaceae; g__Sutterella; s__:359809
3                                          k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Lachnospiraceae; g__; s__:975306
4                                                         k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__; g__; s__:988932
5                                          k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Lachnospiraceae; g__; s__:791522
6                                          k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Lachnospiraceae; g__; s__:368698
7            k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Bifidobacteriales; f__Bifidobacteriaceae; g__Bifidobacterium; s__:553611
8                                        k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Rikenellaceae; g__; s__:953855
9                                          k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Ruminococcaceae; g__; s__:361811
10                                         k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Lachnospiraceae; g__; s__:299858
11 k__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Pasteurellales; f__Pasteurellaceae; g__Haemophilus; s__parainfluenzae:1016422
12                                         k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Lachnospiraceae; g__; s__:564961
13                                        k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Lachnospiraceae; g__; s__:3857426
14                                                        k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__; g__; s__:357471
15                                         k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Ruminococcaceae; g__; s__:544859
      loading
1  0.08771768
2  0.06930874
3  0.06016404
4  0.05532871
5  0.05253113
6  0.03388710
7  0.03278774
8  0.02921690
9  0.02610257
10 0.02557336
11 0.02498921
12 0.02459782
13 0.02406610
14 0.02406610
15 0.02360540</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_top</span>(V_MGX, <span class="st">"MGX"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Top MGX features on PC1:
                                                                                                                                                                      feature
1                                                              k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae|g__Akkermansia
2                                                                                                    k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales
3                                                                             k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae
4                  k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae|g__Akkermansia|s__Akkermansia_muciniphila|t__GCF_000020225
5                                                                                                                          k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae
6                                                                                                                                              k__Bacteria|p__Verrucomicrobia
7                                   k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae|g__Akkermansia|s__Akkermansia_muciniphila
8                                  k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides_sp_4_3_47FAA|t__GCF_000158515
9                                                   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides_sp_4_3_47FAA
10 k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides_thetaiotaomicron|t__Bacteroides_thetaiotaomicron_unclassified
11                                              k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides_thetaiotaomicron
12                                                                                              k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Rikenellaceae
13                                                                                 k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Rikenellaceae|g__Alistipes
14                                             k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Subdoligranulum|s__Subdoligranulum_unclassified
15                                                                             k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Subdoligranulum
      loading
1  0.08951155
2  0.08951155
3  0.08951155
4  0.08951155
5  0.08951155
6  0.08951155
7  0.08951155
8  0.08943717
9  0.08943717
10 0.08802469
11 0.08802469
12 0.08479602
13 0.08479602
14 0.08305904
15 0.08305904</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_top</span>(V_MTX, <span class="st">"MTX"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Top MTX features on PC1:
                                                                                                                                                                      feature
1                                                                                                                          k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae
2                                                                                                    k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales
3                                                                             k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae
4                  k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae|g__Akkermansia|s__Akkermansia_muciniphila|t__GCF_000020225
5                                                                                                                                              k__Bacteria|p__Verrucomicrobia
6                                                              k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae|g__Akkermansia
7                                   k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Verrucomicrobiaceae|g__Akkermansia|s__Akkermansia_muciniphila
8  k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides_thetaiotaomicron|t__Bacteroides_thetaiotaomicron_unclassified
9                                               k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides_thetaiotaomicron
10                                                k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Virgaviridae|g__Tobamovirus|s__Pepper_mild_mottle_virus
11                                  k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Virgaviridae|g__Tobamovirus|s__Pepper_mild_mottle_virus|t__PRJNA15148
12                                                                                              k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Rikenellaceae
13                                                                                 k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Rikenellaceae|g__Alistipes
14                                             k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Subdoligranulum|s__Subdoligranulum_unclassified
15                                                                             k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Subdoligranulum
      loading
1  0.08951155
2  0.08951155
3  0.08951155
4  0.08951155
5  0.08951155
6  0.08951155
7  0.08951155
8  0.08802469
9  0.08802469
10 0.08496505
11 0.08496505
12 0.08479602
13 0.08479602
14 0.08305904
15 0.08305904</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">#variance explained (scree plot)</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>var_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">PC =</span> <span class="fu">names</span>(fit3<span class="sc">$</span>ord.res<span class="sc">$</span>eigvals),</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">Variance =</span> fit3<span class="sc">$</span>ord.res<span class="sc">$</span>eigvals<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> <span class="fu">sum</span>(fit3<span class="sc">$</span>ord.res<span class="sc">$</span>eigvals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(var_df, ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> PC, <span class="at">y =</span> Variance)) <span class="sc">+</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_col</span>(<span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_text</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">label =</span> scales<span class="sc">::</span><span class="fu">percent</span>(Variance, <span class="at">accuracy =</span> <span class="fl">0.1</span>)),</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>                     <span class="at">vjust =</span> <span class="sc">-</span><span class="fl">0.4</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Variance Explained per Component"</span>,</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">x =</span> <span class="st">"Principal Component"</span>, <span class="at">y =</span> <span class="st">"Proportion of Variance"</span>) <span class="sc">+</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ibdmdb_3omic_jointrpca_files/figure-html/setup-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">#cross-view correlation heatmap</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>U_mgx <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(fit3<span class="sc">$</span>ord.res<span class="sc">$</span>samples)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>corr <span class="ot">&lt;-</span> <span class="fu">cor</span>(U_mgx)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(corr, <span class="at">main =</span> <span class="st">"Component correlation matrix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ibdmdb_3omic_jointrpca_files/figure-html/setup-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Session info:</span><span class="sc">\n</span><span class="st">"</span>); <span class="fu">print</span>(<span class="fu">sessionInfo</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Session info:</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.1 (2022-06-23 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 22631)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

attached base packages:
[1] splines   stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] CCA_1.2.2                      fields_15.2                   
 [3] viridisLite_0.4.2              spam_2.10-0                   
 [5] fda_6.3.0                      deSolve_1.40                  
 [7] fds_1.8                        RCurl_1.98-1.14               
 [9] rainbow_3.8                    pcaPP_2.0-4                   
[11] MASS_7.3-57                    knitr_1.50                    
[13] data.table_1.15.4              NMF_0.28                      
[15] cluster_2.1.4                  rngtools_1.5.2                
[17] registry_0.5-1                 ranger_0.16.0                 
[19] tibble_3.2.1                   phyloseq_1.42.0               
[21] HMP2Data_1.12.0                ExperimentHub_2.6.0           
[23] AnnotationHub_3.6.0            BiocFileCache_2.6.1           
[25] dbplyr_2.5.0                   MOFA2_1.8.0                   
[27] randomForest_4.7-1.1           pROC_1.18.5                   
[29] reshape2_1.4.4                 mia_1.17.6                    
[31] TreeSummarizedExperiment_2.6.0 Biostrings_2.66.0             
[33] XVector_0.38.0                 SingleCellExperiment_1.20.1   
[35] MultiAssayExperiment_1.24.0    SummarizedExperiment_1.28.0   
[37] Biobase_2.58.0                 GenomicRanges_1.50.2          
[39] GenomeInfoDb_1.34.9            IRanges_2.32.0                
[41] S4Vectors_0.36.2               BiocGenerics_0.44.0           
[43] MatrixGenerics_1.10.0          matrixStats_1.3.0             
[45] concaveman_1.1.0               ggforce_0.4.2                 
[47] ROCR_1.0-11                    vegan_2.6-4                   
[49] permute_0.9-8                  factoextra_1.0.7              
[51] plotly_4.11.0                  stringr_1.5.1                 
[53] readr_2.1.5                    tidyr_1.3.1                   
[55] dplyr_1.1.4                    caret_7.0-1                   
[57] lattice_0.22-6                 ggplot2_3.5.2                 
[59] glmnet_4.1-8                   optimx_2025-4.9               
[61] pracma_2.4.4                   RSpectra_0.16-1               
[63] irlba_2.3.5.1                  Matrix_1.5-1                  

loaded via a namespace (and not attached):
  [1] rappdirs_0.3.3                ModelMetrics_1.2.2.2         
  [3] bit64_4.0.5                   DelayedArray_0.23.2          
  [5] rpart_4.1.23                  doParallel_1.0.17            
  [7] KEGGREST_1.38.0               hardhat_1.4.1                
  [9] generics_0.1.4                ScaledMatrix_1.6.0           
 [11] cowplot_1.2.0                 RSQLite_2.3.6                
 [13] future_1.29.0                 bit_4.0.5                    
 [15] tzdb_0.4.0                    xml2_1.3.6                   
 [17] lubridate_1.9.3               httpuv_1.6.6                 
 [19] assertthat_0.2.1              DirichletMultinomial_1.40.0  
 [21] viridis_0.6.5                 gower_1.0.1                  
 [23] xfun_0.52                     hms_1.1.3                    
 [25] evaluate_1.0.4                promises_1.3.0               
 [27] igraph_2.0.3                  DBI_1.2.3                    
 [29] htmlwidgets_1.6.4             purrr_1.0.2                  
 [31] ellipsis_0.3.2                ks_1.14.2                    
 [33] corrplot_0.95                 gridBase_0.4-7               
 [35] RcppParallel_5.1.7            sparseMatrixStats_1.10.0     
 [37] vctrs_0.6.5                   cachem_1.0.8                 
 [39] withr_3.0.2                   treeio_1.22.0                
 [41] mclust_6.1                    svglite_2.1.3                
 [43] BiocBaseUtils_1.0.0           dotCall64_1.1-1              
 [45] ape_5.8                       dir.expiry_1.6.0             
 [47] lazyeval_0.2.2                crayon_1.5.3                 
 [49] basilisk.utils_1.10.0         labeling_0.4.3               
 [51] recipes_1.3.1                 pkgconfig_2.0.3              
 [53] slam_0.1-50                   tweenr_2.0.3                 
 [55] nlme_3.1-157                  vipor_0.4.7                  
 [57] nnet_7.3-17                   rlang_1.1.3                  
 [59] globals_0.18.0                lifecycle_1.0.4              
 [61] filelock_1.0.3                rsvd_1.0.5                   
 [63] dichromat_2.0-0.1             polyclip_1.10-6              
 [65] Rhdf5lib_1.20.0               beeswarm_0.4.0               
 [67] pheatmap_1.0.13               png_0.1-8                    
 [69] bitops_1.0-7                  KernSmooth_2.23-20           
 [71] rhdf5filters_1.10.1           blob_1.2.4                   
 [73] DelayedMatrixStats_1.20.0     shape_1.4.6.1                
 [75] decontam_1.18.0               parallelly_1.37.1            
 [77] DECIPHER_2.26.0               beachmat_2.14.2              
 [79] scales_1.4.0                  memoise_2.0.1                
 [81] magrittr_2.0.3                plyr_1.8.9                   
 [83] zlibbioc_1.44.0               hdrcde_3.4                   
 [85] compiler_4.2.1                kableExtra_1.4.0             
 [87] RColorBrewer_1.1-3            cli_3.6.2                    
 [89] ade4_1.7-22                   listenv_0.9.1                
 [91] mgcv_1.8-40                   tidyselect_1.2.1             
 [93] stringi_1.7.8                 forcats_1.0.0                
 [95] yaml_2.3.5                    BiocSingular_1.14.0          
 [97] ggrepel_0.9.5                 grid_4.2.1                   
 [99] tools_4.2.1                   timechange_0.3.0             
[101] future.apply_1.10.0           parallel_4.2.1               
[103] rstudioapi_0.14               bluster_1.8.0                
[105] foreach_1.5.2                 gridExtra_2.3                
[107] prodlim_2023.08.28            farver_2.1.1                 
[109] Rtsne_0.17                    digest_0.6.35                
[111] BiocManager_1.30.26           shiny_1.7.3                  
[113] lava_1.7.0                    Rcpp_1.0.12                  
[115] scuttle_1.8.4                 BiocVersion_3.16.0           
[117] later_1.3.2                   httr_1.4.7                   
[119] AnnotationDbi_1.60.2          colorspace_2.1-0             
[121] fs_1.6.3                      reticulate_1.36.1            
[123] uwot_0.2.2                    yulab.utils_0.2.0            
[125] tidytree_0.4.6                scater_1.26.1                
[127] basilisk_1.10.2               multtest_2.54.0              
[129] systemfonts_1.0.6             xtable_1.8-4                 
[131] jsonlite_1.8.8                nloptr_2.0.3                 
[133] timeDate_4041.110             ipred_0.9-14                 
[135] R6_2.5.1                      pillar_1.11.0                
[137] htmltools_0.5.8.1             mime_0.12                    
[139] glue_1.7.0                    fastmap_1.1.1                
[141] BiocParallel_1.32.6           BiocNeighbors_1.16.0         
[143] class_7.3-20                  interactiveDisplayBase_1.36.0
[145] codetools_0.2-18              maps_3.4.2                   
[147] mvtnorm_1.2-4                 rbiom_1.0.3                  
[149] numDeriv_2016.8-1.1           curl_5.2.1                   
[151] ggbeeswarm_0.7.2              survival_3.6-4               
[153] rmarkdown_2.29                biomformat_1.26.0            
[155] rhdf5_2.42.1                  GenomeInfoDbData_1.2.9       
[157] iterators_1.0.14              HDF5Array_1.26.0             
[159] gtable_0.3.6                 </code></pre>
</div>
</div>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>