<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>IBDMDB — 2-Omic Joint-RPCA (MGX + MTX)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="ibdmdb_2omic_jointrpca_files/libs/clipboard/clipboard.min.js"></script>
<script src="ibdmdb_2omic_jointrpca_files/libs/quarto-html/quarto.js"></script>
<script src="ibdmdb_2omic_jointrpca_files/libs/quarto-html/popper.min.js"></script>
<script src="ibdmdb_2omic_jointrpca_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ibdmdb_2omic_jointrpca_files/libs/quarto-html/anchor.min.js"></script>
<link href="ibdmdb_2omic_jointrpca_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ibdmdb_2omic_jointrpca_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ibdmdb_2omic_jointrpca_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ibdmdb_2omic_jointrpca_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ibdmdb_2omic_jointrpca_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">IBDMDB — 2-Omic Joint-RPCA (MGX + MTX)</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">warn =</span> <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Load user-defined functions</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/dependencies.R"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointRPCA.R"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointRPCAuniversal.R"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointOptspaceHelper.R"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointOptspaceSolve.R"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/optspaceHelper.R"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/transformHelper.R"</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/transform.R"</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/maskValueOnly.R"</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/rpcaTableProcessing.R"</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/jointRPCAutils.R"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr, <span class="at">warn.conflicts =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co">#paths</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>f_mgx <span class="ot">&lt;-</span> <span class="st">"data_ibdmdb_raw/taxonomic_profiles_mgx.tsv"</span>     <span class="co">#metagenomics</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>f_mtx <span class="ot">&lt;-</span> <span class="st">"data_ibdmdb_raw/ecs_relab.tsv"</span>                  <span class="co">#metatranscriptomics</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>f_meta <span class="ot">&lt;-</span> <span class="st">"data_ibdmdb_raw/hmp2_metadata_2018-08-20.csv"</span>  <span class="co">#HMP2/IBDMDB master metadata</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Helper functions</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="co">#read an IBDMDB-style TSV with commented header line(s) beginning with "#"</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>read_ibdmdb_tsv <span class="ot">&lt;-</span> <span class="cf">function</span>(path) {</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stopifnot</span>(<span class="fu">file.exists</span>(path))</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>  first <span class="ot">&lt;-</span> <span class="fu">readLines</span>(path, <span class="at">n =</span> 200L, <span class="at">warn =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>  comment_idx <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">grepl</span>(<span class="st">"^#"</span>, first))</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(comment_idx) <span class="sc">==</span> 0L) <span class="fu">stop</span>(<span class="st">"No commented header line found in: "</span>, path)</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>  header_line <span class="ot">&lt;-</span> first[<span class="fu">max</span>(comment_idx)]</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>  header_line <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"^#</span><span class="sc">\\</span><span class="st">s*"</span>, <span class="st">""</span>, header_line)</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>  header_line <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"^\ufeff"</span>, <span class="st">""</span>, header_line)               </span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>  header_vec  <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(header_line, <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">fixed =</span> <span class="cn">TRUE</span>)[[<span class="dv">1</span>]]</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>  header_vec  <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">'^"|"$'</span>, <span class="st">""</span>, header_vec)                 </span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>  dt <span class="ot">&lt;-</span> <span class="fu">fread</span>(path, <span class="at">skip =</span> <span class="fu">length</span>(comment_idx), <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">ncol</span>(dt) <span class="sc">!=</span> <span class="fu">length</span>(header_vec)) {</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="fu">sprintf</span>(<span class="st">"Header columns (%d) != data columns (%d) in %s"</span>,</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">length</span>(header_vec), <span class="fu">ncol</span>(dt), path))</span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(dt, header_vec)</span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>  dt</span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a><span class="co">#convert data.table (first column = feature IDs) =&gt; numeric matrix (features x samples)</span></span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a>to_matrix <span class="ot">&lt;-</span> <span class="cf">function</span>(dt) {</span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a>  rn <span class="ot">&lt;-</span> dt[[<span class="dv">1</span>]]</span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(dt[, <span class="sc">-</span><span class="dv">1</span>, <span class="at">with =</span> <span class="cn">FALSE</span>])</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(mat) <span class="ot">&lt;-</span> rn</span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">storage.mode</span>(mat) <span class="ot">&lt;-</span> <span class="st">"numeric"</span></span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a>  mat[<span class="fu">is.na</span>(mat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a>  mat</span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a><span class="co">#make rownames unique and ensure numeric / finite entries</span></span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a>dedup_rownames <span class="ot">&lt;-</span> <span class="cf">function</span>(mat) {</span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stopifnot</span>(<span class="sc">!</span><span class="fu">is.null</span>(<span class="fu">rownames</span>(mat)))</span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a>  rn <span class="ot">&lt;-</span> <span class="fu">rownames</span>(mat)</span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a>  rn[rn <span class="sc">==</span> <span class="st">""</span> <span class="sc">|</span> <span class="fu">is.na</span>(rn)] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"feat_"</span>, <span class="fu">seq_len</span>(<span class="fu">sum</span>(rn <span class="sc">==</span> <span class="st">""</span> <span class="sc">|</span> <span class="fu">is.na</span>(rn))))</span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(mat) <span class="ot">&lt;-</span> <span class="fu">make.unique</span>(<span class="fu">as.character</span>(rn), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb1-61"><a href="#cb1-61" aria-hidden="true" tabindex="-1"></a>  mat</span>
<span id="cb1-62"><a href="#cb1-62" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-63"><a href="#cb1-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-64"><a href="#cb1-64" aria-hidden="true" tabindex="-1"></a>sanitize_matrix <span class="ot">&lt;-</span> <span class="cf">function</span>(mat) {</span>
<span id="cb1-65"><a href="#cb1-65" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(mat)</span>
<span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a>  <span class="fu">storage.mode</span>(mat) <span class="ot">&lt;-</span> <span class="st">"numeric"</span></span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a>  mat[<span class="sc">!</span><span class="fu">is.finite</span>(mat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb1-68"><a href="#cb1-68" aria-hidden="true" tabindex="-1"></a>  mat</span>
<span id="cb1-69"><a href="#cb1-69" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-70"><a href="#cb1-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-71"><a href="#cb1-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-72"><a href="#cb1-72" aria-hidden="true" tabindex="-1"></a><span class="co">#try to get an IBD vs non-IBD grouping from metadata</span></span>
<span id="cb1-73"><a href="#cb1-73" aria-hidden="true" tabindex="-1"></a>make_group_from_meta <span class="ot">&lt;-</span> <span class="cf">function</span>(meta_df, sample_ids) {</span>
<span id="cb1-74"><a href="#cb1-74" aria-hidden="true" tabindex="-1"></a>  <span class="co">#always return a frame with sample_id and Group</span></span>
<span id="cb1-75"><a href="#cb1-75" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample_id =</span> sample_ids, <span class="at">Group =</span> <span class="cn">NA_character_</span>, </span>
<span id="cb1-76"><a href="#cb1-76" aria-hidden="true" tabindex="-1"></a>                    <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-77"><a href="#cb1-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-78"><a href="#cb1-78" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(meta_df) <span class="sc">||</span> <span class="sc">!</span><span class="fu">nrow</span>(meta_df)) <span class="fu">return</span>(out)</span>
<span id="cb1-79"><a href="#cb1-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-80"><a href="#cb1-80" aria-hidden="true" tabindex="-1"></a>  <span class="co">#make sure we have a plain data.frame with original names preserved</span></span>
<span id="cb1-81"><a href="#cb1-81" aria-hidden="true" tabindex="-1"></a>  meta_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(meta_df, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-82"><a href="#cb1-82" aria-hidden="true" tabindex="-1"></a>  <span class="co">#standardize names (lowercase + trim) for matching</span></span>
<span id="cb1-83"><a href="#cb1-83" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(meta_df) <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">names</span>(meta_df)))</span>
<span id="cb1-84"><a href="#cb1-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-85"><a href="#cb1-85" aria-hidden="true" tabindex="-1"></a>  <span class="co">#candidate ID columns (lowercase)</span></span>
<span id="cb1-86"><a href="#cb1-86" aria-hidden="true" tabindex="-1"></a>  id_candidates <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">c</span>(</span>
<span id="cb1-87"><a href="#cb1-87" aria-hidden="true" tabindex="-1"></a>    <span class="st">"external id"</span>,<span class="st">"external.id"</span>,<span class="st">"external_id"</span>,</span>
<span id="cb1-88"><a href="#cb1-88" aria-hidden="true" tabindex="-1"></a>    <span class="st">"sample id"</span>,<span class="st">"sample.id"</span>,<span class="st">"sample_id"</span>,</span>
<span id="cb1-89"><a href="#cb1-89" aria-hidden="true" tabindex="-1"></a>    <span class="st">"mgx.sampleid"</span>,<span class="st">"mtx.sampleid"</span>,</span>
<span id="cb1-90"><a href="#cb1-90" aria-hidden="true" tabindex="-1"></a>    <span class="st">"stool sample id:  tube a  (etoh)"</span>,<span class="st">"sample id: tube b (no preservative)"</span>,</span>
<span id="cb1-91"><a href="#cb1-91" aria-hidden="true" tabindex="-1"></a>    <span class="st">"stool_id"</span>,<span class="st">"wr id"</span>,<span class="st">"wr_id"</span>,<span class="st">"wrid"</span></span>
<span id="cb1-92"><a href="#cb1-92" aria-hidden="true" tabindex="-1"></a>  )))</span>
<span id="cb1-93"><a href="#cb1-93" aria-hidden="true" tabindex="-1"></a>  id_col <span class="ot">&lt;-</span> <span class="fu">intersect</span>(id_candidates, <span class="fu">names</span>(meta_df))</span>
<span id="cb1-94"><a href="#cb1-94" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">length</span>(id_col)) <span class="fu">return</span>(out)</span>
<span id="cb1-95"><a href="#cb1-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-96"><a href="#cb1-96" aria-hidden="true" tabindex="-1"></a>  <span class="co">#pick the first ID column that overlaps enough with our sample IDs</span></span>
<span id="cb1-97"><a href="#cb1-97" aria-hidden="true" tabindex="-1"></a>  chosen_id <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb1-98"><a href="#cb1-98" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (cand <span class="cf">in</span> id_col) {</span>
<span id="cb1-99"><a href="#cb1-99" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">as.character</span>(meta_df[[cand]])</span>
<span id="cb1-100"><a href="#cb1-100" aria-hidden="true" tabindex="-1"></a>    hits <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(x) <span class="sc">&amp;</span> x <span class="sc">%in%</span> sample_ids)</span>
<span id="cb1-101"><a href="#cb1-101" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (hits <span class="sc">&gt;=</span> <span class="fu">max</span>(10L, <span class="fu">floor</span>(<span class="fu">length</span>(sample_ids) <span class="sc">*</span> <span class="fl">0.05</span>))) { chosen_id <span class="ot">&lt;-</span> cand; <span class="cf">break</span> }</span>
<span id="cb1-102"><a href="#cb1-102" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-103"><a href="#cb1-103" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(chosen_id)) <span class="fu">return</span>(out)</span>
<span id="cb1-104"><a href="#cb1-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-105"><a href="#cb1-105" aria-hidden="true" tabindex="-1"></a>  md <span class="ot">&lt;-</span> meta_df[, <span class="fu">c</span>(chosen_id, <span class="fu">setdiff</span>(<span class="fu">names</span>(meta_df), chosen_id)), drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb1-106"><a href="#cb1-106" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(md)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"sample_id"</span></span>
<span id="cb1-107"><a href="#cb1-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-108"><a href="#cb1-108" aria-hidden="true" tabindex="-1"></a>  <span class="co">#candidate diagnosis columns</span></span>
<span id="cb1-109"><a href="#cb1-109" aria-hidden="true" tabindex="-1"></a>  diag_candidates <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">c</span>(</span>
<span id="cb1-110"><a href="#cb1-110" aria-hidden="true" tabindex="-1"></a>    <span class="st">"diagnosis"</span>, <span class="st">"dx"</span>, <span class="st">"disease_status"</span>, <span class="st">"disease.status"</span>,</span>
<span id="cb1-111"><a href="#cb1-111" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ibd_status"</span>, <span class="st">"ibd.status"</span>, <span class="st">"ibd_subtype"</span>, <span class="st">"diagnosis_subtype"</span></span>
<span id="cb1-112"><a href="#cb1-112" aria-hidden="true" tabindex="-1"></a>  )))</span>
<span id="cb1-113"><a href="#cb1-113" aria-hidden="true" tabindex="-1"></a>  dcols <span class="ot">&lt;-</span> <span class="fu">intersect</span>(diag_candidates, <span class="fu">names</span>(md))</span>
<span id="cb1-114"><a href="#cb1-114" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">length</span>(dcols)) <span class="fu">return</span>(out)</span>
<span id="cb1-115"><a href="#cb1-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-116"><a href="#cb1-116" aria-hidden="true" tabindex="-1"></a>  <span class="co">#try columns until one produces enough labels</span></span>
<span id="cb1-117"><a href="#cb1-117" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (col <span class="cf">in</span> dcols) {</span>
<span id="cb1-118"><a href="#cb1-118" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">as.character</span>(md[[col]])</span>
<span id="cb1-119"><a href="#cb1-119" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">length</span>(x) <span class="sc">||</span> <span class="fu">length</span>(x) <span class="sc">!=</span> <span class="fu">nrow</span>(md)) <span class="cf">next</span></span>
<span id="cb1-120"><a href="#cb1-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-121"><a href="#cb1-121" aria-hidden="true" tabindex="-1"></a>    grp <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA_character_</span>, <span class="fu">nrow</span>(md))</span>
<span id="cb1-122"><a href="#cb1-122" aria-hidden="true" tabindex="-1"></a>    is_ibd  <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">b(uc|cd|ibd)</span><span class="sc">\\</span><span class="st">b"</span>, x, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-123"><a href="#cb1-123" aria-hidden="true" tabindex="-1"></a>    is_non  <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"^</span><span class="sc">\\</span><span class="st">s*non"</span>, x, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-124"><a href="#cb1-124" aria-hidden="true" tabindex="-1"></a>    grp[is_ibd] <span class="ot">&lt;-</span> <span class="st">"IBD"</span></span>
<span id="cb1-125"><a href="#cb1-125" aria-hidden="true" tabindex="-1"></a>    grp[is_non <span class="sc">&amp;</span> <span class="sc">!</span>is_ibd] <span class="ot">&lt;-</span> <span class="st">"non-IBD"</span></span>
<span id="cb1-126"><a href="#cb1-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-127"><a href="#cb1-127" aria-hidden="true" tabindex="-1"></a>    n_lab <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(grp))</span>
<span id="cb1-128"><a href="#cb1-128" aria-hidden="true" tabindex="-1"></a>    n_cls <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">na.omit</span>(grp)))</span>
<span id="cb1-129"><a href="#cb1-129" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (n_lab <span class="sc">&gt;=</span> <span class="fu">max</span>(10L, <span class="fu">floor</span>(<span class="fu">length</span>(sample_ids) <span class="sc">*</span> <span class="fl">0.05</span>)) <span class="sc">&amp;&amp;</span> n_cls <span class="sc">&gt;=</span> <span class="dv">2</span>) {</span>
<span id="cb1-130"><a href="#cb1-130" aria-hidden="true" tabindex="-1"></a>      md<span class="sc">$</span>Group <span class="ot">&lt;-</span> grp</span>
<span id="cb1-131"><a href="#cb1-131" aria-hidden="true" tabindex="-1"></a>      joined <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb1-132"><a href="#cb1-132" aria-hidden="true" tabindex="-1"></a>        out,</span>
<span id="cb1-133"><a href="#cb1-133" aria-hidden="true" tabindex="-1"></a>        md[, <span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"Group"</span>)],</span>
<span id="cb1-134"><a href="#cb1-134" aria-hidden="true" tabindex="-1"></a>        <span class="at">by =</span> <span class="st">"sample_id"</span></span>
<span id="cb1-135"><a href="#cb1-135" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb1-136"><a href="#cb1-136" aria-hidden="true" tabindex="-1"></a>      joined<span class="sc">$</span>Group <span class="ot">&lt;-</span> <span class="fu">factor</span>(joined<span class="sc">$</span>Group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"IBD"</span>, <span class="st">"non-IBD"</span>))</span>
<span id="cb1-137"><a href="#cb1-137" aria-hidden="true" tabindex="-1"></a>      <span class="fu">return</span>(joined)</span>
<span id="cb1-138"><a href="#cb1-138" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb1-139"><a href="#cb1-139" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-140"><a href="#cb1-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-141"><a href="#cb1-141" aria-hidden="true" tabindex="-1"></a>  <span class="co">#if nothing usable, just return NAs</span></span>
<span id="cb1-142"><a href="#cb1-142" aria-hidden="true" tabindex="-1"></a>  out</span>
<span id="cb1-143"><a href="#cb1-143" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-144"><a href="#cb1-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-145"><a href="#cb1-145" aria-hidden="true" tabindex="-1"></a><span class="co">#evaluate V1..V3 scores with Wilcoxon, PERMANOVA, weighted RF AUROC</span></span>
<span id="cb1-146"><a href="#cb1-146" aria-hidden="true" tabindex="-1"></a>eval_scores <span class="ot">&lt;-</span> <span class="cf">function</span>(scores_df) {</span>
<span id="cb1-147"><a href="#cb1-147" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb1-148"><a href="#cb1-148" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span>(<span class="st">"Group"</span> <span class="sc">%in%</span> <span class="fu">names</span>(scores_df))) <span class="fu">return</span>(out)</span>
<span id="cb1-149"><a href="#cb1-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-150"><a href="#cb1-150" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Wilcoxon PC1/PC2</span></span>
<span id="cb1-151"><a href="#cb1-151" aria-hidden="true" tabindex="-1"></a>  res_w1 <span class="ot">&lt;-</span> <span class="fu">try</span>(<span class="fu">wilcox.test</span>(scores_df<span class="sc">$</span>V1 <span class="sc">~</span> scores_df<span class="sc">$</span>Group, <span class="at">exact =</span> <span class="cn">FALSE</span>), <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-152"><a href="#cb1-152" aria-hidden="true" tabindex="-1"></a>  res_w2 <span class="ot">&lt;-</span> <span class="fu">try</span>(<span class="fu">wilcox.test</span>(scores_df<span class="sc">$</span>V2 <span class="sc">~</span> scores_df<span class="sc">$</span>Group, <span class="at">exact =</span> <span class="cn">FALSE</span>), <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-153"><a href="#cb1-153" aria-hidden="true" tabindex="-1"></a>  out<span class="sc">$</span>wilcox_PC1_p <span class="ot">&lt;-</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">inherits</span>(res_w1, <span class="st">"try-error"</span>)) res_w1<span class="sc">$</span>p.value <span class="cf">else</span> <span class="cn">NA_real_</span></span>
<span id="cb1-154"><a href="#cb1-154" aria-hidden="true" tabindex="-1"></a>  out<span class="sc">$</span>wilcox_PC2_p <span class="ot">&lt;-</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">inherits</span>(res_w2, <span class="st">"try-error"</span>)) res_w2<span class="sc">$</span>p.value <span class="cf">else</span> <span class="cn">NA_real_</span></span>
<span id="cb1-155"><a href="#cb1-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-156"><a href="#cb1-156" aria-hidden="true" tabindex="-1"></a>  <span class="co">#PERMANOVA on V1..V3</span></span>
<span id="cb1-157"><a href="#cb1-157" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">all</span>(<span class="fu">c</span>(<span class="st">"V1"</span>, <span class="st">"V2"</span>, <span class="st">"V3"</span>) <span class="sc">%in%</span> <span class="fu">names</span>(scores_df))) {</span>
<span id="cb1-158"><a href="#cb1-158" aria-hidden="true" tabindex="-1"></a>    perm_df <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(scores_df[, <span class="fu">c</span>(<span class="st">"Group"</span>, <span class="st">"V1"</span>, <span class="st">"V2"</span>, <span class="st">"V3"</span>)])</span>
<span id="cb1-159"><a href="#cb1-159" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">is.factor</span>(perm_df<span class="sc">$</span>Group) <span class="sc">&amp;&amp;</span> <span class="fu">nlevels</span>(perm_df<span class="sc">$</span>Group) <span class="sc">&gt;=</span> <span class="dv">2</span> <span class="sc">&amp;&amp;</span> <span class="fu">all</span>(<span class="fu">table</span>(perm_df<span class="sc">$</span>Group) <span class="sc">&gt;=</span> <span class="dv">5</span>)) {</span>
<span id="cb1-160"><a href="#cb1-160" aria-hidden="true" tabindex="-1"></a>      perm <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(perm_df[, <span class="fu">c</span>(<span class="st">"V1"</span>,<span class="st">"V2"</span>,<span class="st">"V3"</span>)] <span class="sc">~</span> Group, <span class="at">data =</span> perm_df, <span class="at">method =</span> <span class="st">"euclidean"</span>)</span>
<span id="cb1-161"><a href="#cb1-161" aria-hidden="true" tabindex="-1"></a>      out<span class="sc">$</span>permanova_R2 <span class="ot">&lt;-</span> perm<span class="sc">$</span>R2[<span class="dv">1</span>]; out<span class="sc">$</span>permanova_F <span class="ot">&lt;-</span> perm<span class="sc">$</span>F[<span class="dv">1</span>]; out<span class="sc">$</span>permanova_p <span class="ot">&lt;-</span> perm<span class="sc">$</span><span class="st">`</span><span class="at">Pr(&gt;F)</span><span class="st">`</span>[<span class="dv">1</span>]</span>
<span id="cb1-162"><a href="#cb1-162" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb1-163"><a href="#cb1-163" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-164"><a href="#cb1-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-165"><a href="#cb1-165" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Weighted RF AUROC</span></span>
<span id="cb1-166"><a href="#cb1-166" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">all</span>(<span class="fu">c</span>(<span class="st">"V1"</span>, <span class="st">"V2"</span>, <span class="st">"V3"</span>) <span class="sc">%in%</span> <span class="fu">names</span>(scores_df))) {</span>
<span id="cb1-167"><a href="#cb1-167" aria-hidden="true" tabindex="-1"></a>    rf_df <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(scores_df[, <span class="fu">c</span>(<span class="st">"Group"</span>, <span class="st">"V1"</span>, <span class="st">"V2"</span>, <span class="st">"V3"</span>)])</span>
<span id="cb1-168"><a href="#cb1-168" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">is.factor</span>(rf_df<span class="sc">$</span>Group) <span class="sc">&amp;&amp;</span> <span class="fu">nlevels</span>(rf_df<span class="sc">$</span>Group) <span class="sc">&gt;=</span> <span class="dv">2</span>) {</span>
<span id="cb1-169"><a href="#cb1-169" aria-hidden="true" tabindex="-1"></a>      cls_tab <span class="ot">&lt;-</span> <span class="fu">table</span>(rf_df<span class="sc">$</span>Group); wts <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="dv">1</span><span class="sc">/</span>cls_tab); <span class="fu">names</span>(wts) <span class="ot">&lt;-</span> <span class="fu">names</span>(cls_tab)</span>
<span id="cb1-170"><a href="#cb1-170" aria-hidden="true" tabindex="-1"></a>      <span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb1-171"><a href="#cb1-171" aria-hidden="true" tabindex="-1"></a>      rf_prob <span class="ot">&lt;-</span> <span class="fu">ranger</span>(Group <span class="sc">~</span> ., <span class="at">data =</span> rf_df, <span class="at">num.trees =</span> <span class="dv">1000</span>,</span>
<span id="cb1-172"><a href="#cb1-172" aria-hidden="true" tabindex="-1"></a>                        <span class="at">probability =</span> <span class="cn">TRUE</span>, <span class="at">class.weights =</span> wts, <span class="at">oob.error =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-173"><a href="#cb1-173" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="st">"IBD"</span> <span class="sc">%in%</span> <span class="fu">colnames</span>(rf_prob<span class="sc">$</span>predictions)) {</span>
<span id="cb1-174"><a href="#cb1-174" aria-hidden="true" tabindex="-1"></a>        p_ibd <span class="ot">&lt;-</span> rf_prob<span class="sc">$</span>predictions[, <span class="st">"IBD"</span>]</span>
<span id="cb1-175"><a href="#cb1-175" aria-hidden="true" tabindex="-1"></a>        roc_obj <span class="ot">&lt;-</span> pROC<span class="sc">::</span><span class="fu">roc</span>(rf_df<span class="sc">$</span>Group, p_ibd, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"non-IBD"</span>, <span class="st">"IBD"</span>))</span>
<span id="cb1-176"><a href="#cb1-176" aria-hidden="true" tabindex="-1"></a>        out<span class="sc">$</span>AUROC <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(pROC<span class="sc">::</span><span class="fu">auc</span>(roc_obj))</span>
<span id="cb1-177"><a href="#cb1-177" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb1-178"><a href="#cb1-178" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb1-179"><a href="#cb1-179" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-180"><a href="#cb1-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-181"><a href="#cb1-181" aria-hidden="true" tabindex="-1"></a>  out</span>
<span id="cb1-182"><a href="#cb1-182" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-183"><a href="#cb1-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-184"><a href="#cb1-184" aria-hidden="true" tabindex="-1"></a><span class="co"># Load MGX / MTX / VRX and build matrices</span></span>
<span id="cb1-185"><a href="#cb1-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-186"><a href="#cb1-186" aria-hidden="true" tabindex="-1"></a>dt_mgx <span class="ot">&lt;-</span> <span class="fu">read_ibdmdb_tsv</span>(f_mgx)</span>
<span id="cb1-187"><a href="#cb1-187" aria-hidden="true" tabindex="-1"></a>dt_mtx <span class="ot">&lt;-</span> <span class="fu">read_ibdmdb_tsv</span>(f_mtx)</span>
<span id="cb1-188"><a href="#cb1-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-189"><a href="#cb1-189" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Dims (rows, cols):</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Dims (rows, cols):</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"MGX:"</span>, <span class="fu">dim</span>(dt_mgx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MGX: 1479 1639 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"MTX:"</span>, <span class="fu">dim</span>(dt_mtx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MTX: 70711 736 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>mat_mgx <span class="ot">&lt;-</span> <span class="fu">to_matrix</span>(dt_mgx)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>mat_mtx <span class="ot">&lt;-</span> <span class="fu">to_matrix</span>(dt_mtx)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#harmonize samples</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>shared_2 <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(mat_mgx), <span class="fu">colnames</span>(mat_mtx))</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>shared_2 <span class="ot">&lt;-</span> <span class="fu">unique</span>(shared_2)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>shared_2 <span class="ot">&lt;-</span> shared_2[<span class="fu">nchar</span>(shared_2) <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Shared MGX–MTX samples:"</span>, <span class="fu">length</span>(shared_2), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Shared MGX–MTX samples: 734 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">length</span>(shared_2) <span class="sc">&gt;=</span> <span class="dv">20</span>) <span class="co">#sanity threshold</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>shared_2 <span class="ot">&lt;-</span> <span class="fu">sort</span>(shared_2)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> mat_mgx[, shared_2, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> mat_mtx[, shared_2, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">#guard against duplicate or empty sample IDs</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="sc">!</span><span class="fu">any</span>(<span class="fu">is.na</span>(shared_2)), <span class="fu">all</span>(<span class="fu">nchar</span>(shared_2) <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">any</span>(<span class="fu">duplicated</span>(shared_2))) {</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">"[guard] Duplicated sample IDs detected; making them unique."</span>)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  new_ids <span class="ot">&lt;-</span> <span class="fu">make.unique</span>(shared_2, <span class="at">sep =</span> <span class="st">"_dup"</span>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(X_mgx) <span class="ot">&lt;-</span> new_ids</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(X_mtx) <span class="ot">&lt;-</span> new_ids</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  shared_2 <span class="ot">&lt;-</span> new_ids</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Per-modality adaptive filtering</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>n_samp <span class="ot">&lt;-</span> <span class="fu">length</span>(shared_2)</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>prev_mgx <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fl">0.05</span> <span class="sc">*</span> n_samp) <span class="co">#5% prevalence</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>prev_mtx <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fl">0.02</span> <span class="sc">*</span> n_samp) <span class="co">#2% prevalence</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>keep_mgx <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(X_mgx <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> prev_mgx</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>keep_mtx <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(X_mtx <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> prev_mtx</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> X_mgx[keep_mgx, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> X_mtx[keep_mtx, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a><span class="co">#drop samples that are all-zero in a given view </span></span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>keep_samp_mgx <span class="ot">&lt;-</span> <span class="fu">colSums</span>(X_mgx) <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>keep_samp_mtx <span class="ot">&lt;-</span> <span class="fu">colSums</span>(X_mtx) <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">all</span>(keep_samp_mgx)) <span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[MGX] dropping %d all-zero samples"</span>, <span class="fu">sum</span>(<span class="sc">!</span>keep_samp_mgx)))</span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">all</span>(keep_samp_mtx)) <span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[MTX] dropping %d all-zero samples"</span>, <span class="fu">sum</span>(<span class="sc">!</span>keep_samp_mtx)))</span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> X_mgx[, keep_samp_mgx, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> X_mtx[, keep_samp_mtx, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a><span class="co">#recompute shared samples after filtering &amp; pruning</span></span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a>shared_final <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(X_mgx), <span class="fu">colnames</span>(X_mtx))</span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>shared_final <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(shared_final))</span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">length</span>(shared_final) <span class="sc">&gt;=</span> <span class="dv">20</span>)  </span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-42"><a href="#cb9-42" aria-hidden="true" tabindex="-1"></a><span class="co">#subset both views to the same (final) sample set and identical order</span></span>
<span id="cb9-43"><a href="#cb9-43" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> X_mgx[, shared_final, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb9-44"><a href="#cb9-44" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> X_mtx[, shared_final, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb9-45"><a href="#cb9-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-46"><a href="#cb9-46" aria-hidden="true" tabindex="-1"></a><span class="co">#final sanity on dimensions</span></span>
<span id="cb9-47"><a href="#cb9-47" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">nrow</span>(X_mgx) <span class="sc">==</span> 0L <span class="sc">||</span> <span class="fu">nrow</span>(X_mtx) <span class="sc">==</span> 0L)</span>
<span id="cb9-48"><a href="#cb9-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"After filtering, one view has zero features: MGX rows="</span>, <span class="fu">nrow</span>(X_mgx), <span class="st">", MTX rows="</span>, <span class="fu">nrow</span>(X_mtx))</span>
<span id="cb9-49"><a href="#cb9-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">ncol</span>(X_mgx) <span class="sc">==</span> 0L <span class="sc">||</span> <span class="fu">ncol</span>(X_mtx) <span class="sc">==</span> 0L)</span>
<span id="cb9-50"><a href="#cb9-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"After filtering, there are zero shared samples."</span>)</span>
<span id="cb9-51"><a href="#cb9-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-52"><a href="#cb9-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Cap MTX by variance for speed</span></span>
<span id="cb9-53"><a href="#cb9-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-54"><a href="#cb9-54" aria-hidden="true" tabindex="-1"></a>max_mtx_features <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb9-55"><a href="#cb9-55" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">nrow</span>(X_mtx) <span class="sc">&gt;</span> max_mtx_features) {</span>
<span id="cb9-56"><a href="#cb9-56" aria-hidden="true" tabindex="-1"></a>mtx_var <span class="ot">&lt;-</span> matrixStats<span class="sc">::</span><span class="fu">rowVars</span>(X_mtx)</span>
<span id="cb9-57"><a href="#cb9-57" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">order</span>(mtx_var, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-58"><a href="#cb9-58" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> X_mtx[ord[<span class="fu">seq_len</span>(max_mtx_features)], , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb9-59"><a href="#cb9-59" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[MTX] Kept top %d features by variance."</span>, max_mtx_features))</span>
<span id="cb9-60"><a href="#cb9-60" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-61"><a href="#cb9-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-62"><a href="#cb9-62" aria-hidden="true" tabindex="-1"></a><span class="co"># Clean rownames / numeric</span></span>
<span id="cb9-63"><a href="#cb9-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-64"><a href="#cb9-64" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> <span class="fu">sanitize_matrix</span>(<span class="fu">dedup_rownames</span>(X_mgx))</span>
<span id="cb9-65"><a href="#cb9-65" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> <span class="fu">sanitize_matrix</span>(<span class="fu">dedup_rownames</span>(X_mtx))</span>
<span id="cb9-66"><a href="#cb9-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-67"><a href="#cb9-67" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Remaining features after filtering:</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb9-68"><a href="#cb9-68" aria-hidden="true" tabindex="-1"></a><span class="st">"MGX:"</span>, <span class="fu">nrow</span>(X_mgx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb9-69"><a href="#cb9-69" aria-hidden="true" tabindex="-1"></a><span class="st">"MTX:"</span>, <span class="fu">nrow</span>(X_mtx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Remaining features after filtering:
 MGX: 459 
 MTX: 10000 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Final dims — MGX:"</span>, <span class="fu">nrow</span>(X_mgx), <span class="st">"x"</span>, <span class="fu">ncol</span>(X_mgx),</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"| MTX:"</span>, <span class="fu">nrow</span>(X_mtx), <span class="st">"x"</span>, <span class="fu">ncol</span>(X_mtx), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Final dims — MGX: 459 x 732 | MTX: 10000 x 732 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">identical</span>(<span class="fu">colnames</span>(X_mgx), <span class="fu">colnames</span>(X_mtx)))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Build MAE from the final sample set and run Joint-RPCA</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co">#colData must match the final column set exactly</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>cd <span class="ot">&lt;-</span> S4Vectors<span class="sc">::</span><span class="fu">DataFrame</span>(<span class="at">row.names =</span> <span class="fu">colnames</span>(X_mgx))   </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co">#SummarizedExperiments (rows = features, cols = samples)</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>se_mgx <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">SummarizedExperiment</span>(</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays  =</span> <span class="fu">list</span>(<span class="at">counts =</span> X_mgx),</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> cd</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>se_mtx <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">SummarizedExperiment</span>(</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays  =</span> <span class="fu">list</span>(<span class="at">counts =</span> X_mtx),</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> cd</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a><span class="co">#MultiAssayExperiment </span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>mae_2 <span class="ot">&lt;-</span> MultiAssayExperiment<span class="sc">::</span><span class="fu">MultiAssayExperiment</span>(</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">experiments =</span> <span class="fu">list</span>(<span class="at">MGX =</span> se_mgx, <span class="at">MTX =</span> se_mtx)</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>mae_2 <span class="ot">&lt;-</span> MultiAssayExperiment<span class="sc">::</span><span class="fu">intersectColumns</span>(mae_2)</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a><span class="co">#choose a safe rank k</span></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>per_view_min_dim <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">list</span>(X_mgx, X_mtx), <span class="cf">function</span>(m) <span class="fu">min</span>(<span class="fu">nrow</span>(m), <span class="fu">ncol</span>(m)))</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>k_max <span class="ot">&lt;-</span> <span class="fu">max</span>(1L, <span class="fu">min</span>(per_view_min_dim))</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">min</span>(3L, k_max)</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[2-omic] per-view min dims = %s; using k = %d"</span>,</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>                <span class="fu">paste</span>(per_view_min_dim, <span class="at">collapse =</span> <span class="st">", "</span>), k))</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>fit2 <span class="ot">&lt;-</span> <span class="fu">jointRPCAuniversal</span>(</span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> mae_2,</span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">n.components =</span> k,</span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">max.iterations =</span> <span class="dv">500</span>,</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">rclr.transform.tables =</span> <span class="cn">TRUE</span>,</span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.sample.count =</span> <span class="dv">1</span>,</span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.feature.count =</span> <span class="dv">0</span>,</span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.feature.frequency =</span> <span class="dv">0</span></span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true" tabindex="-1"></a><span class="co">#inspect structure so we don’t assume a specific layout for features</span></span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(fit2<span class="sc">$</span>ord.res, <span class="at">max.level =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 7
 $ method              : chr "rpca"
 $ eigvals             : Named num [1:3] 550 489 461
  ..- attr(*, "names")= chr [1:3] "PC1" "PC2" "PC3"
 $ samples             : num [1:732, 1:3] -0.0397 0.0229 -0.0268 -0.0323 -0.0122 ...
  ..- attr(*, "dimnames")=List of 2
 $ features            : num [1:10459, 1:3] -0.000878 -0.000878 -0.000878 -0.000878 -0.000878 ...
  ..- attr(*, "dimnames")=List of 2
 $ proportion.explained: Named num [1:3] 0.401 0.317 0.282
  ..- attr(*, "names")= chr [1:3] "PC1" "PC2" "PC3"
 $ dist                : NULL
 $ metadata            : list()
 - attr(*, "class")= chr "OrdinationResults"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#extract sample scores (U)</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>U <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(fit2<span class="sc">$</span>ord.res<span class="sc">$</span>samples)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(U) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"V"</span>, <span class="fu">seq_len</span>(<span class="fu">ncol</span>(U)))</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>U<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">rownames</span>(U)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co">#extract per-modality loadings robustly</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co">#some builds store ord.res$features as a list per assay; others as a single object</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>get_view <span class="ot">&lt;-</span> <span class="cf">function</span>(obj, keys) {</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(obj)) <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.list</span>(obj)) {</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (k <span class="cf">in</span> keys) <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(obj[[k]])) <span class="fu">return</span>(<span class="fu">as.data.frame</span>(obj[[k]]))</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  <span class="co">#fallback: if it's a matrix/data.frame but not split by view, skip splitting</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.matrix</span>(obj) <span class="sc">||</span> <span class="fu">is.data.frame</span>(obj)) <span class="fu">return</span>(<span class="fu">as.data.frame</span>(obj))</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>V_mgx <span class="ot">&lt;-</span> <span class="fu">get_view</span>(fit2<span class="sc">$</span>ord.res<span class="sc">$</span>features, <span class="fu">c</span>(<span class="st">"MGX"</span>,<span class="st">"view_MGX"</span>, <span class="st">"view_mgx"</span>))</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>V_mtx <span class="ot">&lt;-</span> <span class="fu">get_view</span>(fit2<span class="sc">$</span>ord.res<span class="sc">$</span>features, <span class="fu">c</span>(<span class="st">"MTX"</span>,<span class="st">"view_MTX"</span>, <span class="st">"view_mtx"</span>))</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="co">#plot ordination (PC1 vs PC2, unlabeled for now)</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(U, <span class="fu">aes</span>(V1, V2)) <span class="sc">+</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">sprintf</span>(<span class="st">"Joint-RPCA (MGX + MTX), k = %d — PC1 vs PC2"</span>, k),</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>         <span class="at">x =</span> <span class="st">"PC1"</span>, <span class="at">y =</span> <span class="st">"PC2"</span>) <span class="sc">+</span></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ibdmdb_2omic_jointrpca_files/figure-html/setup-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>make_groups_autodetect <span class="ot">&lt;-</span> <span class="cf">function</span>(meta_df, sample_ids, <span class="at">min_frac =</span> <span class="fl">0.01</span>, <span class="at">min_abs =</span> 10L) {</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">#always return a data.frame(sample_id, Group)</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample_id =</span> sample_ids, <span class="at">Group =</span> <span class="fu">factor</span>(<span class="cn">NA</span>, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"IBD"</span>, <span class="st">"non-IBD"</span>)),</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(meta_df) <span class="sc">||</span> <span class="sc">!</span><span class="fu">nrow</span>(meta_df)) <span class="fu">return</span>(out)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  md <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(meta_df, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(md) <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">names</span>(md)))</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>  sid <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(sample_ids)))</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  thresh <span class="ot">&lt;-</span> <span class="fu">max</span>(min_abs, <span class="fu">floor</span>(<span class="fu">length</span>(sid) <span class="sc">*</span> min_frac))</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#overlap of every column with our sample IDs</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>  overlaps <span class="ot">&lt;-</span> <span class="fu">sapply</span>(md, <span class="cf">function</span>(col) {</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(col)))</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(x) <span class="sc">&amp;</span> x <span class="sc">%in%</span> sid)</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>  <span class="co">#choose the best column, but only if it clears the threshold</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>  max_ov <span class="ot">&lt;-</span> <span class="fu">suppressWarnings</span>(<span class="fu">max</span>(overlaps, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.finite</span>(max_ov) <span class="sc">||</span> max_ov <span class="sc">&lt;</span> thresh) {</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[meta] No metadata column matches sample IDs (max overlap = %s &lt; %d). Group left NA."</span>,</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">as.character</span>(max_ov), thresh))</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(out)</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>  best <span class="ot">&lt;-</span> <span class="fu">names</span>(overlaps)[<span class="fu">which.max</span>(overlaps)]</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"[meta] Chosen sample-id column: '%s' (matches = %d)"</span>, best, overlaps[[best]]))</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="st">"diagnosis"</span> <span class="sc">%in%</span> <span class="fu">names</span>(md)) {</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="st">"[meta] 'diagnosis' not found; Group left NA."</span>)</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(out)</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a>  <span class="co">#build IBD / non-IBD from diagnosis</span></span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a>  dx  <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(md<span class="sc">$</span>diagnosis)))</span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>  grp <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">b(uc|cd|ibd)</span><span class="sc">\\</span><span class="st">b"</span>, dx), <span class="st">"IBD"</span>,</span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>         <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"^</span><span class="sc">\\</span><span class="st">s*non"</span>, dx), <span class="st">"non-IBD"</span>, <span class="cn">NA_character_</span>))</span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>  md<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">trimws</span>(<span class="fu">as.character</span>(md[[best]])))</span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>  md<span class="sc">$</span>Group <span class="ot">&lt;-</span> <span class="fu">factor</span>(grp, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"IBD"</span>, <span class="st">"non-IBD"</span>))</span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a>  join_tbl <span class="ot">&lt;-</span> <span class="fu">unique</span>(md[, <span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"Group"</span>)])</span>
<span id="cb16-43"><a href="#cb16-43" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb16-44"><a href="#cb16-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(<span class="at">sample_id =</span> sid, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>),</span>
<span id="cb16-45"><a href="#cb16-45" aria-hidden="true" tabindex="-1"></a>    join_tbl,</span>
<span id="cb16-46"><a href="#cb16-46" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"sample_id"</span></span>
<span id="cb16-47"><a href="#cb16-47" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb16-48"><a href="#cb16-48" aria-hidden="true" tabindex="-1"></a>  out<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> sample_ids  <span class="co">#restore original casing</span></span>
<span id="cb16-49"><a href="#cb16-49" aria-hidden="true" tabindex="-1"></a>  out</span>
<span id="cb16-50"><a href="#cb16-50" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-51"><a href="#cb16-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-52"><a href="#cb16-52" aria-hidden="true" tabindex="-1"></a>meta_df <span class="ot">&lt;-</span> <span class="cf">if</span> (<span class="fu">file.exists</span>(f_meta)) <span class="fu">data.frame</span>(data.table<span class="sc">::</span><span class="fu">fread</span>(f_meta)) <span class="cf">else</span> <span class="cn">NULL</span></span>
<span id="cb16-53"><a href="#cb16-53" aria-hidden="true" tabindex="-1"></a>grp_df  <span class="ot">&lt;-</span> <span class="fu">make_groups_autodetect</span>(meta_df, U<span class="sc">$</span>sample_id)</span>
<span id="cb16-54"><a href="#cb16-54" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">sort</span>(<span class="fu">table</span>(grp_df<span class="sc">$</span>Group, <span class="at">useNA =</span> <span class="st">"ifany"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
non-IBD     IBD 
    187     545 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>U2 <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(U, grp_df, <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co">#color the ordination by IBD status</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(U2, <span class="fu">aes</span>(V1, V2, <span class="at">color =</span> Group)) <span class="sc">+</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">size =</span> <span class="fl">1.2</span>) <span class="sc">+</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">sprintf</span>(<span class="st">"Joint-RPCA (k=%d): PC1 vs PC2 by Group"</span>, k)) <span class="sc">+</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ibdmdb_2omic_jointrpca_files/figure-html/setup-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">#basic stats: Wilcoxon on PC1/PC2, PERMANOVA on PC1–PC3, RF AUROC</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> <span class="fu">eval_scores</span>(U2)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(stats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$wilcox_PC1_p
[1] 0.03047366

$wilcox_PC2_p
[1] 2.168776e-09

$permanova_R2
[1] 0.01724885

$permanova_F
[1] 12.81267

$permanova_p
[1] 0.001

$AUROC
[1] 0.7418535</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#see how many samples couldn’t be labeled</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(U2<span class="sc">$</span>Group))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#quick top-loading features (only if per-view loadings were available)</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>top_k <span class="ot">&lt;-</span> <span class="dv">15</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(V_mgx) <span class="sc">&amp;&amp;</span> <span class="fu">ncol</span>(V_mgx) <span class="sc">&gt;=</span> <span class="dv">1</span>) {</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  ord <span class="ot">&lt;-</span> <span class="fu">order</span>(V_mgx[,<span class="dv">1</span>], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Top MGX features on PC1:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">data.frame</span>(</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature =</span> <span class="fu">rownames</span>(V_mgx)[ord][<span class="fu">seq_len</span>(<span class="fu">min</span>(top_k, <span class="fu">length</span>(ord)))],</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">loading =</span> V_mgx[ord, <span class="dv">1</span>][<span class="fu">seq_len</span>(<span class="fu">min</span>(top_k, <span class="fu">length</span>(ord)))]</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Top MGX features on PC1:
                                                                                                                  feature
1                                                      2.7.1.69: Protein-N(pi)-phosphohistidine--sugar phosphotransferase
2  2.7.1.69: Protein-N(pi)-phosphohistidine--sugar phosphotransferase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
3                                2.7.7.6: DNA-directed RNA polymerase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
4                       3.6.3.14: H(+)-transporting two-sector ATPase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
5                                                                              4.2.1.55: 3-hydroxybutyryl-CoA dehydratase
6                          4.2.1.55: 3-hydroxybutyryl-CoA dehydratase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
7                   2.7.7.27: Glucose-1-phosphate adenylyltransferase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
8                                1.3.99.2: Transferred entry: 1.3.8.1|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
9    2.7.1.90: Diphosphate--fructose-6-phosphate 1-phosphotransferase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
10                                                                                   1.3.99.2: Transferred entry: 1.3.8.1
11                5.3.1.17: 5-dehydro-4-deoxy-D-glucuronate isomerase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
12                                                                              1.1.1.35: 3-hydroxyacyl-CoA dehydrogenase
13                                                                                2.3.1.9: Acetyl-CoA C-acetyltransferase
14                          1.1.1.35: 3-hydroxyacyl-CoA dehydrogenase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
15                      1.1.1.157: 3-hydroxybutyryl-CoA dehydrogenase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
      loading
1  0.06607556
2  0.06158873
3  0.06150761
4  0.06008987
5  0.05708062
6  0.05622509
7  0.05539683
8  0.05405219
9  0.05397505
10 0.05396237
11 0.05355906
12 0.05308296
13 0.05289232
14 0.05265166
15 0.05265166</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(V_mtx) <span class="sc">&amp;&amp;</span> <span class="fu">ncol</span>(V_mtx) <span class="sc">&gt;=</span> <span class="dv">1</span>) {</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  ord <span class="ot">&lt;-</span> <span class="fu">order</span>(V_mtx[,<span class="dv">1</span>], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Top MTX features on PC1:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">data.frame</span>(</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature =</span> <span class="fu">rownames</span>(V_mtx)[ord][<span class="fu">seq_len</span>(<span class="fu">min</span>(top_k, <span class="fu">length</span>(ord)))],</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">loading =</span> V_mtx[ord, <span class="dv">1</span>][<span class="fu">seq_len</span>(<span class="fu">min</span>(top_k, <span class="fu">length</span>(ord)))]</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Top MTX features on PC1:
                                                                                                                  feature
1                                                      2.7.1.69: Protein-N(pi)-phosphohistidine--sugar phosphotransferase
2  2.7.1.69: Protein-N(pi)-phosphohistidine--sugar phosphotransferase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
3                                2.7.7.6: DNA-directed RNA polymerase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
4                       3.6.3.14: H(+)-transporting two-sector ATPase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
5                                                                              4.2.1.55: 3-hydroxybutyryl-CoA dehydratase
6                          4.2.1.55: 3-hydroxybutyryl-CoA dehydratase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
7                   2.7.7.27: Glucose-1-phosphate adenylyltransferase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
8                                1.3.99.2: Transferred entry: 1.3.8.1|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
9    2.7.1.90: Diphosphate--fructose-6-phosphate 1-phosphotransferase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
10                                                                                   1.3.99.2: Transferred entry: 1.3.8.1
11                5.3.1.17: 5-dehydro-4-deoxy-D-glucuronate isomerase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
12                                                                              1.1.1.35: 3-hydroxyacyl-CoA dehydrogenase
13                                                                                2.3.1.9: Acetyl-CoA C-acetyltransferase
14                          1.1.1.35: 3-hydroxyacyl-CoA dehydrogenase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
15                      1.1.1.157: 3-hydroxybutyryl-CoA dehydrogenase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
      loading
1  0.06607556
2  0.06158873
3  0.06150761
4  0.06008987
5  0.05708062
6  0.05622509
7  0.05539683
8  0.05405219
9  0.05397505
10 0.05396237
11 0.05355906
12 0.05308296
13 0.05289232
14 0.05265166
15 0.05265166</code></pre>
</div>
</div>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>