<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>IBDMDB — 2-Omic Joint-RPCA (MGX + MTX)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="ibdmdb_2omic_jointrpca-mia_demo_files/libs/clipboard/clipboard.min.js"></script>
<script src="ibdmdb_2omic_jointrpca-mia_demo_files/libs/quarto-html/quarto.js"></script>
<script src="ibdmdb_2omic_jointrpca-mia_demo_files/libs/quarto-html/popper.min.js"></script>
<script src="ibdmdb_2omic_jointrpca-mia_demo_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ibdmdb_2omic_jointrpca-mia_demo_files/libs/quarto-html/anchor.min.js"></script>
<link href="ibdmdb_2omic_jointrpca-mia_demo_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ibdmdb_2omic_jointrpca-mia_demo_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ibdmdb_2omic_jointrpca-mia_demo_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ibdmdb_2omic_jointrpca-mia_demo_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ibdmdb_2omic_jointrpca-mia_demo_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">IBDMDB — 2-Omic Joint-RPCA (MGX + MTX)</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">warn =</span> <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Install merged branch containing Joint-RPCA</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"microbiome/mia"</span>, <span class="at">ref =</span> <span class="st">"rpca"</span>, <span class="at">upgrade =</span> <span class="st">"never"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co">#necessary imports</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Helpers</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>.get_view <span class="ot">&lt;-</span> <span class="cf">function</span>(obj, keys) {</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(obj)) <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.list</span>(obj)) {</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (k <span class="cf">in</span> keys) <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(obj[[k]])) <span class="fu">return</span>(<span class="fu">as.data.frame</span>(obj[[k]]))</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>  <span class="co">#fallback: if it's a matrix/data.frame but not split by view, skip splitting</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.matrix</span>(obj) <span class="sc">||</span> <span class="fu">is.data.frame</span>(obj)) <span class="fu">return</span>(<span class="fu">as.data.frame</span>(obj))</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>.show_top <span class="ot">&lt;-</span> <span class="cf">function</span>(V, label, <span class="at">top_k =</span> <span class="dv">15</span>) {</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(V) <span class="sc">||</span> <span class="sc">!</span><span class="fu">ncol</span>(V)) <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>  ord <span class="ot">&lt;-</span> <span class="fu">order</span>(V[, <span class="dv">1</span>], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>  top_ix <span class="ot">&lt;-</span> <span class="fu">seq_len</span>(<span class="fu">min</span>(top_k, <span class="fu">length</span>(ord)))</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">label =</span> label,</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature =</span> <span class="fu">rownames</span>(V)[ord][top_ix],</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">loading =</span> V[ord, <span class="dv">1</span>][top_ix],</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">row.names =</span> <span class="cn">NULL</span>,</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a><span class="co">#load packaged demo data</span></span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(ibdmdb_2omic_demo)</span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a><span class="co"># MultiAssayExperiment with MGX + MTX</span></span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a>mae_2 <span class="ot">&lt;-</span> ibdmdb_2omic_demo</span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a><span class="co">#extract assays (features x samples)</span></span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>se_mgx <span class="ot">&lt;-</span> mae_2[[<span class="st">"MGX"</span>]]</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>se_mtx <span class="ot">&lt;-</span> mae_2[[<span class="st">"MTX"</span>]]</span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a>X_mgx <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">assay</span>(se_mgx, 1L)</span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a>X_mtx <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">assay</span>(se_mtx, 1L)</span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">identical</span>(<span class="fu">colnames</span>(X_mgx), <span class="fu">colnames</span>(X_mtx)))</span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a>sample_ids <span class="ot">&lt;-</span> <span class="fu">colnames</span>(X_mgx)</span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a><span class="co">#choose a safe rank k and run Joint-RPCA</span></span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-61"><a href="#cb1-61" aria-hidden="true" tabindex="-1"></a>per_view_min_dim <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">list</span>(X_mgx, X_mtx), <span class="cf">function</span>(m) <span class="fu">min</span>(<span class="fu">nrow</span>(m), <span class="fu">ncol</span>(m)))</span>
<span id="cb1-62"><a href="#cb1-62" aria-hidden="true" tabindex="-1"></a>k_max <span class="ot">&lt;-</span> <span class="fu">max</span>(1L, <span class="fu">min</span>(per_view_min_dim))</span>
<span id="cb1-63"><a href="#cb1-63" aria-hidden="true" tabindex="-1"></a>k     <span class="ot">&lt;-</span> <span class="fu">min</span>(3L, k_max)</span>
<span id="cb1-64"><a href="#cb1-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-65"><a href="#cb1-65" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">sprintf</span>(</span>
<span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a>    <span class="st">"[2-omic] per-view min dims = %s; using k = %d"</span>,</span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(per_view_min_dim, <span class="at">collapse =</span> <span class="st">", "</span>), k</span>
<span id="cb1-68"><a href="#cb1-68" aria-hidden="true" tabindex="-1"></a>))</span>
<span id="cb1-69"><a href="#cb1-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-70"><a href="#cb1-70" aria-hidden="true" tabindex="-1"></a><span class="co"># make sure both experiments have an assay called "counts" </span></span>
<span id="cb1-71"><a href="#cb1-71" aria-hidden="true" tabindex="-1"></a>se_mgx <span class="ot">&lt;-</span> mae_2[[<span class="st">"MGX"</span>]]</span>
<span id="cb1-72"><a href="#cb1-72" aria-hidden="true" tabindex="-1"></a>se_mtx <span class="ot">&lt;-</span> mae_2[[<span class="st">"MTX"</span>]]</span>
<span id="cb1-73"><a href="#cb1-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-74"><a href="#cb1-74" aria-hidden="true" tabindex="-1"></a>SummarizedExperiment<span class="sc">::</span><span class="fu">assay</span>(se_mgx, <span class="st">"counts"</span>) <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">assay</span>(se_mgx, 1L)</span>
<span id="cb1-75"><a href="#cb1-75" aria-hidden="true" tabindex="-1"></a>SummarizedExperiment<span class="sc">::</span><span class="fu">assay</span>(se_mtx, <span class="st">"counts"</span>) <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">assay</span>(se_mtx, 1L)</span>
<span id="cb1-76"><a href="#cb1-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-77"><a href="#cb1-77" aria-hidden="true" tabindex="-1"></a>mae_2[[<span class="st">"MGX"</span>]] <span class="ot">&lt;-</span> se_mgx</span>
<span id="cb1-78"><a href="#cb1-78" aria-hidden="true" tabindex="-1"></a>mae_2[[<span class="st">"MTX"</span>]] <span class="ot">&lt;-</span> se_mtx</span>
<span id="cb1-79"><a href="#cb1-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-80"><a href="#cb1-80" aria-hidden="true" tabindex="-1"></a><span class="co"># call jointRPCAuniversal</span></span>
<span id="cb1-81"><a href="#cb1-81" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb1-82"><a href="#cb1-82" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> mae_2,</span>
<span id="cb1-83"><a href="#cb1-83" aria-hidden="true" tabindex="-1"></a>  <span class="at">transform =</span> <span class="st">"rclr"</span>,</span>
<span id="cb1-84"><a href="#cb1-84" aria-hidden="true" tabindex="-1"></a>  <span class="at">n.components =</span> k,</span>
<span id="cb1-85"><a href="#cb1-85" aria-hidden="true" tabindex="-1"></a>  <span class="at">max.iterations =</span> <span class="dv">5</span>,</span>
<span id="cb1-86"><a href="#cb1-86" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.sample.count =</span> <span class="dv">0</span>,</span>
<span id="cb1-87"><a href="#cb1-87" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.feature.count =</span> <span class="dv">0</span>,</span>
<span id="cb1-88"><a href="#cb1-88" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.feature.frequency =</span> <span class="dv">0</span></span>
<span id="cb1-89"><a href="#cb1-89" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-90"><a href="#cb1-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-91"><a href="#cb1-91" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> args[<span class="fu">names</span>(args) <span class="sc">%in%</span> <span class="fu">names</span>(<span class="fu">formals</span>(mia<span class="sc">:::</span>jointRPCAuniversal))]</span>
<span id="cb1-92"><a href="#cb1-92" aria-hidden="true" tabindex="-1"></a>fit2 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(mia<span class="sc">:::</span>jointRPCAuniversal, args)</span>
<span id="cb1-93"><a href="#cb1-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-94"><a href="#cb1-94" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(mae_2)<span class="sc">$</span>JointRPCA <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">JointRPCA =</span> fit2)</span>
<span id="cb1-95"><a href="#cb1-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-96"><a href="#cb1-96" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb1-97"><a href="#cb1-97" aria-hidden="true" tabindex="-1"></a><span class="co">#scores and loadings</span></span>
<span id="cb1-98"><a href="#cb1-98" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb1-99"><a href="#cb1-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-100"><a href="#cb1-100" aria-hidden="true" tabindex="-1"></a>U <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(fit2<span class="sc">$</span>ord_res<span class="sc">$</span>samples)</span>
<span id="cb1-101"><a href="#cb1-101" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(U) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"V"</span>, <span class="fu">seq_len</span>(<span class="fu">ncol</span>(U)))</span>
<span id="cb1-102"><a href="#cb1-102" aria-hidden="true" tabindex="-1"></a>U<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">trimws</span>(<span class="fu">rownames</span>(U))</span>
<span id="cb1-103"><a href="#cb1-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-104"><a href="#cb1-104" aria-hidden="true" tabindex="-1"></a>V_mgx <span class="ot">&lt;-</span> <span class="fu">.get_view</span>(fit2<span class="sc">$</span>ord_res<span class="sc">$</span>features, <span class="fu">c</span>(<span class="st">"MGX"</span>, <span class="st">"view_MGX"</span>, <span class="st">"view_mgx"</span>))</span>
<span id="cb1-105"><a href="#cb1-105" aria-hidden="true" tabindex="-1"></a>V_mtx <span class="ot">&lt;-</span> <span class="fu">.get_view</span>(fit2<span class="sc">$</span>ord_res<span class="sc">$</span>features, <span class="fu">c</span>(<span class="st">"MTX"</span>, <span class="st">"view_MTX"</span>, <span class="st">"view_mtx"</span>))</span>
<span id="cb1-106"><a href="#cb1-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-107"><a href="#cb1-107" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb1-108"><a href="#cb1-108" aria-hidden="true" tabindex="-1"></a><span class="co">#attach IBD/non-IBD groups and plot ordination</span></span>
<span id="cb1-109"><a href="#cb1-109" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb1-110"><a href="#cb1-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-111"><a href="#cb1-111" aria-hidden="true" tabindex="-1"></a>meta_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(SummarizedExperiment<span class="sc">::</span><span class="fu">colData</span>(mae_2), <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-112"><a href="#cb1-112" aria-hidden="true" tabindex="-1"></a>meta_df<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">rownames</span>(meta_df)</span>
<span id="cb1-113"><a href="#cb1-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-114"><a href="#cb1-114" aria-hidden="true" tabindex="-1"></a>meta_df<span class="sc">$</span>diagnosis <span class="ot">&lt;-</span> <span class="fu">as.character</span>(meta_df<span class="sc">$</span>diagnosis)</span>
<span id="cb1-115"><a href="#cb1-115" aria-hidden="true" tabindex="-1"></a>meta_df<span class="sc">$</span>Group <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(</span>
<span id="cb1-116"><a href="#cb1-116" aria-hidden="true" tabindex="-1"></a>  meta_df<span class="sc">$</span>diagnosis <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"UC"</span>, <span class="st">"CD"</span>) <span class="sc">~</span> <span class="st">"IBD"</span>,</span>
<span id="cb1-117"><a href="#cb1-117" aria-hidden="true" tabindex="-1"></a>  <span class="fu">grepl</span>(<span class="st">"^non"</span>, meta_df<span class="sc">$</span>diagnosis, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>) <span class="sc">~</span> <span class="st">"non-IBD"</span>,</span>
<span id="cb1-118"><a href="#cb1-118" aria-hidden="true" tabindex="-1"></a>  <span class="cn">TRUE</span> <span class="sc">~</span> <span class="cn">NA_character_</span></span>
<span id="cb1-119"><a href="#cb1-119" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-120"><a href="#cb1-120" aria-hidden="true" tabindex="-1"></a>meta_df<span class="sc">$</span>Group <span class="ot">&lt;-</span> <span class="fu">factor</span>(meta_df<span class="sc">$</span>Group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"IBD"</span>, <span class="st">"non-IBD"</span>))</span>
<span id="cb1-121"><a href="#cb1-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-122"><a href="#cb1-122" aria-hidden="true" tabindex="-1"></a>grp_df <span class="ot">&lt;-</span> meta_df[, <span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"Group"</span>)]</span>
<span id="cb1-123"><a href="#cb1-123" aria-hidden="true" tabindex="-1"></a>grp_df<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">as.character</span>(grp_df<span class="sc">$</span>sample_id)</span>
<span id="cb1-124"><a href="#cb1-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-125"><a href="#cb1-125" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">sort</span>(<span class="fu">table</span>(grp_df<span class="sc">$</span>Group, <span class="at">useNA =</span> <span class="st">"ifany"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
non-IBD     IBD 
     15      45 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>U2 <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(U, grp_df, <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>U2_plot <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(U2, <span class="sc">!</span><span class="fu">is.na</span>(Group))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>tab <span class="ot">&lt;-</span> <span class="fu">table</span>(U2<span class="sc">$</span>Group, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    IBD non-IBD 
     45      15 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#PC1–PC2 colored by Group</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(U2_plot, <span class="fu">aes</span>(<span class="at">x =</span> V1, <span class="at">y =</span> V2, <span class="at">color =</span> Group)) <span class="sc">+</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.85</span>, <span class="at">size =</span> <span class="fl">1.2</span>) <span class="sc">+</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">title =</span> <span class="fu">sprintf</span>(<span class="st">"Joint-RPCA (MGX + MTX), k = %d — PC1 vs PC2"</span>, k),</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> <span class="st">"PC1"</span>, <span class="at">y =</span> <span class="st">"PC2"</span>, <span class="at">color =</span> <span class="cn">NULL</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ibdmdb_2omic_jointrpca-mia_demo_files/figure-html/setup-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co">#quick top-loadings on PC1 for each view</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># --------------------------------------------------------------------</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">.show_top</span>(V_mgx, <span class="st">"MGX"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   label
1    MGX
2    MGX
3    MGX
4    MGX
5    MGX
6    MGX
7    MGX
8    MGX
9    MGX
10   MGX
11   MGX
12   MGX
13   MGX
14   MGX
15   MGX
                                                                                                                                                                             feature
1                                                                                                     k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Prevotellaceae
2                                                                                       k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Prevotellaceae|g__Prevotella
3                                                                 k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Lachnospiraceae|g__Roseburia|s__Roseburia_intestinalis
4                          k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Lachnospiraceae|g__Roseburia|s__Roseburia_intestinalis|t__Roseburia_intestinalis_unclassified
5                                                                   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Prevotellaceae|g__Prevotella|s__Prevotella_copri
6                                                  k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Prevotellaceae|g__Prevotella|s__Prevotella_copri|t__GCF_000157935
7                                                                                           k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Lachnospiraceae|g__Roseburia
8                                                                                                                               k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria
9                                                                                    k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Veillonella
10                                                                                                            k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria|o__Pasteurellales
11                                                                                         k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria|o__Pasteurellales|f__Pasteurellaceae
12                                                                          k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria|o__Pasteurellales|f__Pasteurellaceae|g__Haemophilus
13                                            k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria|o__Pasteurellales|f__Pasteurellaceae|g__Haemophilus|s__Haemophilus_parainfluenzae
14 k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria|o__Pasteurellales|f__Pasteurellaceae|g__Haemophilus|s__Haemophilus_parainfluenzae|t__Haemophilus_parainfluenzae_unclassified
15                                                                                                                                     k__Bacteria|p__Firmicutes|c__Erysipelotrichia
      loading
1  0.11116099
2  0.11054843
3  0.09198188
4  0.09198188
5  0.08061194
6  0.08061194
7  0.07177956
8  0.07139797
9  0.06877179
10 0.06580455
11 0.06580455
12 0.06571709
13 0.06484041
14 0.06484041
15 0.06425434</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">.show_top</span>(V_mtx, <span class="st">"MTX"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   label
1    MTX
2    MTX
3    MTX
4    MTX
5    MTX
6    MTX
7    MTX
8    MTX
9    MTX
10   MTX
11   MTX
12   MTX
13   MTX
14   MTX
15   MTX
                                                                                           feature
1                                                                          1.3.99.1: Deleted entry
2                                                            2.1.1.148: Thymidylate synthase (FAD)
3                                                          2.3.1.9: Acetyl-CoA C-acetyltransferase
4                                                             1.3.99.2: Transferred entry: 1.3.8.1
5                                                            2.3.1.54: Formate C-acetyltransferase
6                                                             1.4.1.13: Glutamate synthase (NADPH)
7                   2.3.1.9: Acetyl-CoA C-acetyltransferase|g__Roseburia.s__Roseburia_intestinalis
8                    2.3.1.9: Acetyl-CoA C-acetyltransferase|g__Eubacterium.s__Eubacterium_rectale
9                                                              1.4.1.14: Glutamate synthase (NADH)
10                                                 1.17.4.2: Ribonucleoside-triphosphate reductase
11 1.1.1.133: dTDP-4-dehydrorhamnose reductase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
12   2.3.1.179: Beta-ketoacyl-[acyl-carrier-protein] synthase II|g__Prevotella.s__Prevotella_copri
13       1.6.5.8: NADH:ubiquinone reductase (Na(+)-transporting)|g__Prevotella.s__Prevotella_copri
14           1.1.1.86: Ketol-acid reductoisomerase (NADP(+))|g__Eubacterium.s__Eubacterium_rectale
15     2.1.3.3: Ornithine carbamoyltransferase|g__Faecalibacterium.s__Faecalibacterium_prausnitzii
      loading
1  0.07137847
2  0.06676899
3  0.05991062
4  0.05884659
5  0.05884553
6  0.05498142
7  0.05488308
8  0.05466168
9  0.05387744
10 0.05386520
11 0.05237605
12 0.05053765
13 0.04999125
14 0.04941186
15 0.04764687</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#label: session-info</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.2 (2025-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 22631)

Matrix products: default
  LAPACK version 3.12.1

locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: Asia/Qyzylorda
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] mia_1.19.2                      TreeSummarizedExperiment_2.18.0
 [3] Biostrings_2.78.0               XVector_0.50.0                 
 [5] SingleCellExperiment_1.32.0     MultiAssayExperiment_1.36.1    
 [7] SummarizedExperiment_1.40.0     Biobase_2.70.0                 
 [9] GenomicRanges_1.62.0            Seqinfo_1.0.0                  
[11] IRanges_2.44.0                  S4Vectors_0.48.0               
[13] BiocGenerics_0.56.0             generics_0.1.4                 
[15] MatrixGenerics_1.22.0           matrixStats_1.5.0              
[17] ggplot2_4.0.1                   tidyr_1.3.1                    
[19] dplyr_1.1.4                    

loaded via a namespace (and not attached):
 [1] DBI_1.2.3                   gridExtra_2.3              
 [3] remotes_2.5.0               permute_0.9-8              
 [5] rlang_1.1.6                 magrittr_2.0.4             
 [7] scater_1.38.0               compiler_4.5.2             
 [9] mgcv_1.9-3                  DelayedMatrixStats_1.32.0  
[11] vctrs_0.6.5                 reshape2_1.4.5             
[13] stringr_1.6.0               pkgconfig_2.0.3            
[15] crayon_1.5.3                fastmap_1.2.0              
[17] scuttle_1.20.0              labeling_0.4.3             
[19] rmarkdown_2.30              ggbeeswarm_0.7.3           
[21] DirichletMultinomial_1.52.0 purrr_1.2.0                
[23] xfun_0.54                   bluster_1.20.0             
[25] beachmat_2.26.0             jsonlite_2.0.0             
[27] DelayedArray_0.36.0         BiocParallel_1.44.0        
[29] irlba_2.3.5.1               parallel_4.5.2             
[31] cluster_2.1.8.1             R6_2.6.1                   
[33] stringi_1.8.7               RColorBrewer_1.1-3         
[35] Rcpp_1.1.0                  knitr_1.50                 
[37] DECIPHER_3.6.0              BiocBaseUtils_1.12.0       
[39] Matrix_1.7-4                splines_4.5.2              
[41] igraph_2.2.1                tidyselect_1.2.1           
[43] rstudioapi_0.17.1           abind_1.4-8                
[45] yaml_2.3.11                 viridis_0.6.5              
[47] vegan_2.7-2                 codetools_0.2-20           
[49] curl_7.0.0                  lattice_0.22-7             
[51] tibble_3.3.0                plyr_1.8.9                 
[53] treeio_1.34.0               withr_3.0.2                
[55] S7_0.2.1                    evaluate_1.0.5             
[57] pillar_1.11.1               sparseMatrixStats_1.22.0   
[59] scales_1.4.0                tidytree_0.4.7             
[61] glue_1.8.0                  lazyeval_0.2.2             
[63] tools_4.5.2                 BiocNeighbors_2.4.0        
[65] ScaledMatrix_1.18.0         fs_1.6.6                   
[67] grid_4.5.2                  ape_5.8-1                  
[69] ecodive_2.2.1               nlme_3.1-168               
[71] beeswarm_0.4.0              BiocSingular_1.26.1        
[73] vipor_0.4.7                 cli_3.6.5                  
[75] rsvd_1.0.5                  rappdirs_0.3.3             
[77] S4Arrays_1.10.1             viridisLite_0.4.2          
[79] gtable_0.3.6                yulab.utils_0.2.3          
[81] digest_0.6.39               SparseArray_1.10.4         
[83] ggrepel_0.9.6               decontam_1.30.0            
[85] htmlwidgets_1.6.4           farver_2.1.2               
[87] htmltools_0.5.9             lifecycle_1.0.5            
[89] MASS_7.3-65                </code></pre>
</div>
</div>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>